// @tldraw/editor/dist-esm/lib/editor/tools/HandTool/children/Dragging.mjs
import { Vec2d } from "@tldraw/primitives";
import { HAND_TOOL_FRICTION } from "../../../../constants.mjs";
import { StateNode } from "../../StateNode.mjs";
class Dragging extends StateNode {
  static id = "dragging";
  onEnter = () => {
    this.update();
  };
  onPointerMove = () => {
    this.update();
  };
  onPointerUp = () => {
    this.complete();
  };
  onCancel = () => {
    this.complete();
  };
  onComplete = () => {
    this.complete();
  };
  update() {
    const { currentScreenPoint, previousScreenPoint } = this.editor.inputs;
    const delta = Vec2d.Sub(currentScreenPoint, previousScreenPoint);
    if (Math.abs(delta.x) > 0 || Math.abs(delta.y) > 0) {
      this.editor.pan(delta.x, delta.y);
    }
  }
  complete() {
    this.editor.slideCamera({
      speed: Math.min(2, this.editor.inputs.pointerVelocity.len()),
      direction: this.editor.inputs.pointerVelocity,
      friction: HAND_TOOL_FRICTION
    });
    this.parent.transition("idle", {});
  }
}
export {
  Dragging
};
