// @tldraw/ui/dist-esm/lib/components/NavigationZone/Minimap.mjs
import { jsx } from "react/jsx-runtime";
import {
  ANIMATION_MEDIUM_MS,
  normalizeWheel,
  useContainer,
  useEditor,
  useQuickReactor
} from "@tldraw/editor";
import { Box2d, Vec2d, intersectPolygonPolygon } from "@tldraw/primitives";
import * as React from "react";
import { track } from "signia-react";
import { MinimapManager } from "./MinimapManager.mjs";
const Minimap = track(function Minimap2({
  shapeFill,
  selectFill,
  viewportFill
}) {
  const editor = useEditor();
  const rCanvas = React.useRef(null);
  const container = useContainer();
  const rPointing = React.useRef(false);
  const minimap = React.useMemo(() => new MinimapManager(editor, editor.devicePixelRatio), [editor]);
  const isDarkMode = editor.isDarkMode;
  React.useEffect(() => {
    const raf = requestAnimationFrame(() => {
      const style = getComputedStyle(container);
      minimap.colors = {
        shapeFill: style.getPropertyValue(shapeFill).trim(),
        selectFill: style.getPropertyValue(selectFill).trim(),
        viewportFill: style.getPropertyValue(viewportFill).trim()
      };
      minimap.render();
    });
    return () => {
      cancelAnimationFrame(raf);
    };
  }, [container, selectFill, shapeFill, viewportFill, minimap, isDarkMode]);
  const onDoubleClick = React.useCallback(
    (e) => {
      if (!editor.currentPageShapeIds.size)
        return;
      const { x, y } = minimap.minimapScreenPointToPagePoint(e.clientX, e.clientY, false, false);
      const clampedPoint = minimap.minimapScreenPointToPagePoint(e.clientX, e.clientY, false, true);
      minimap.originPagePoint.setTo(clampedPoint);
      minimap.originPageCenter.setTo(editor.viewportPageBounds.center);
      editor.centerOnPoint(x, y, { duration: ANIMATION_MEDIUM_MS });
    },
    [editor, minimap]
  );
  const onPointerDown = React.useCallback(
    (e) => {
      e.currentTarget.setPointerCapture(e.pointerId);
      if (!editor.currentPageShapeIds.size)
        return;
      rPointing.current = true;
      minimap.isInViewport = false;
      const { x, y } = minimap.minimapScreenPointToPagePoint(e.clientX, e.clientY, false, false);
      const clampedPoint = minimap.minimapScreenPointToPagePoint(e.clientX, e.clientY, false, true);
      const _vpPageBounds = editor.viewportPageBounds;
      minimap.originPagePoint.setTo(clampedPoint);
      minimap.originPageCenter.setTo(_vpPageBounds.center);
      minimap.isInViewport = _vpPageBounds.containsPoint(clampedPoint);
      if (!minimap.isInViewport) {
        editor.centerOnPoint(x, y, { duration: ANIMATION_MEDIUM_MS });
      }
    },
    [editor, minimap]
  );
  const onPointerMove = React.useCallback(
    (e) => {
      if (rPointing.current) {
        const { x, y } = minimap.minimapScreenPointToPagePoint(
          e.clientX,
          e.clientY,
          e.shiftKey,
          true
        );
        if (minimap.isInViewport) {
          const delta = Vec2d.Sub({ x, y }, minimap.originPagePoint);
          const center = Vec2d.Add(minimap.originPageCenter, delta);
          editor.centerOnPoint(center.x, center.y);
          return;
        }
        editor.centerOnPoint(x, y);
      }
      const pagePoint = minimap.getPagePoint(e.clientX, e.clientY);
      const screenPoint = editor.pageToScreen(pagePoint.x, pagePoint.y);
      const info = {
        type: "pointer",
        target: "canvas",
        name: "pointer_move",
        ...getPointerInfo(e),
        point: screenPoint,
        isPen: editor.isPenMode
      };
      editor.dispatch(info);
    },
    [editor, minimap]
  );
  const onPointerUp = React.useCallback((_e) => {
    rPointing.current = false;
  }, []);
  const onWheel = React.useCallback(
    (e) => {
      const offset = normalizeWheel(e);
      editor.dispatch({
        type: "wheel",
        name: "wheel",
        delta: offset,
        shiftKey: e.shiftKey,
        altKey: e.altKey,
        ctrlKey: e.metaKey || e.ctrlKey
      });
    },
    [editor]
  );
  useQuickReactor(
    "update dpr",
    () => {
      const { devicePixelRatio } = editor;
      minimap.setDpr(devicePixelRatio);
      const canvas = rCanvas.current;
      const rect = canvas.getBoundingClientRect();
      const width = rect.width * devicePixelRatio;
      const height = rect.height * devicePixelRatio;
      canvas.width = width;
      canvas.height = height;
      minimap.canvasScreenBounds.set(rect.x, rect.y, width, height);
      minimap.cvs = rCanvas.current;
    },
    [editor, minimap]
  );
  const presences = React.useMemo(() => {
    return editor.store.query.records("instance_presence");
  }, [editor]);
  useQuickReactor(
    "minimap render when pagebounds or collaborators changes",
    () => {
      const { devicePixelRatio, viewportPageBounds, allShapesCommonBounds } = editor;
      devicePixelRatio;
      minimap.contentPageBounds = allShapesCommonBounds ? Box2d.Expand(allShapesCommonBounds, viewportPageBounds) : viewportPageBounds;
      minimap.updateContentScreenBounds();
      const allShapeBounds = [];
      editor.currentPageShapeIds.forEach((id) => {
        let pageBounds = editor.getPageBoundsById(id);
        const pageMask = editor.getPageMaskById(id);
        if (pageMask) {
          const intersection = intersectPolygonPolygon(pageMask, pageBounds.corners);
          if (!intersection) {
            return;
          }
          pageBounds = Box2d.FromPoints(intersection);
        }
        if (pageBounds) {
          pageBounds.id = id;
          allShapeBounds.push(pageBounds);
        }
      });
      minimap.pageBounds = allShapeBounds;
      minimap.collaborators = presences.value;
      minimap.render();
    },
    [editor, minimap]
  );
  return /* @__PURE__ */ jsx("div", { className: "tlui-minimap", children: /* @__PURE__ */ jsx(
    "canvas",
    {
      ref: rCanvas,
      className: "tlui-minimap__canvas",
      onDoubleClick,
      onPointerMove,
      onPointerDown,
      onPointerUp,
      onWheel
    }
  ) });
});
function getPointerInfo(e) {
  ;
  e.isKilled = true;
  return {
    point: {
      x: e.clientX,
      y: e.clientY,
      z: e.pressure
    },
    shiftKey: e.shiftKey,
    altKey: e.altKey,
    ctrlKey: e.metaKey || e.ctrlKey,
    pointerId: e.pointerId,
    button: e.button,
    isPen: e.pointerType === "pen"
  };
}
export {
  Minimap
};
