// @tldraw/editor/dist-cjs/lib/editor/tools/SelectTool/children/ScribbleBrushing.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ScribbleBrushing_exports = {};
__export(ScribbleBrushing_exports, {
  ScribbleBrushing: () => ScribbleBrushing
});
module.exports = __toCommonJS(ScribbleBrushing_exports);
var import_primitives = require("@tldraw/primitives");
var import_ScribbleManager = require("../../../managers/ScribbleManager");
var import_FrameShapeUtil = require("../../../shapes/frame/FrameShapeUtil");
var import_GroupShapeUtil = require("../../../shapes/group/GroupShapeUtil");
var import_StateNode = require("../../StateNode");
class ScribbleBrushing extends import_StateNode.StateNode {
  static id = "scribble_brushing";
  hits = /* @__PURE__ */ new Set();
  size = 0;
  scribble = {};
  initialSelectedIds = /* @__PURE__ */ new Set();
  newlySelectedIds = /* @__PURE__ */ new Set();
  onEnter = () => {
    this.initialSelectedIds = new Set(
      this.editor.inputs.shiftKey ? this.editor.selectedIds : []
    );
    this.newlySelectedIds = /* @__PURE__ */ new Set();
    this.size = 0;
    this.hits.clear();
    this.startScribble();
    this.updateBrushSelection();
    requestAnimationFrame(() => {
      this.editor.setBrush(null);
    });
  };
  onExit = () => {
    this.editor.setErasingIds([]);
    this.scribble.stop();
  };
  onPointerMove = () => {
    this.updateBrushSelection();
  };
  onPointerUp = () => {
    this.complete();
  };
  onKeyDown = () => {
    this.updateBrushSelection();
  };
  onKeyUp = () => {
    if (!this.editor.inputs.altKey) {
      this.parent.transition("brushing", {});
    } else {
      this.updateBrushSelection();
    }
  };
  startScribble = () => {
    if (this.scribble.tick) {
      this.editor.off("tick", this.scribble?.tick);
    }
    this.scribble = new import_ScribbleManager.ScribbleManager({
      onUpdate: this.onScribbleUpdate,
      onComplete: this.onScribbleComplete,
      color: "selection-stroke",
      opacity: 0.32,
      size: 12
    });
    this.editor.on("tick", this.scribble.tick);
  };
  pushPointToScribble = () => {
    const { x, y } = this.editor.inputs.currentPagePoint;
    this.scribble.addPoint(x, y);
  };
  onScribbleUpdate = (scribble) => {
    this.editor.setScribble(scribble);
  };
  onScribbleComplete = () => {
    this.editor.off("tick", this.scribble.tick);
    this.editor.setScribble(null);
  };
  updateBrushSelection() {
    const {
      shapesArray,
      inputs: { originPagePoint, previousPagePoint, currentPagePoint }
    } = this.editor;
    this.pushPointToScribble();
    const shapes = shapesArray;
    let shape, util;
    for (let i = 0, n = shapes.length; i < n; i++) {
      shape = shapes[i];
      util = this.editor.getShapeUtil(shape);
      if (this.editor.isShapeOfType(shape, import_GroupShapeUtil.GroupShapeUtil) || this.newlySelectedIds.has(shape.id) || this.editor.isShapeOfType(shape, import_FrameShapeUtil.FrameShapeUtil) && util.hitTestPoint(shape, this.editor.getPointInShapeSpace(shape, originPagePoint)) || this.editor.isShapeOrAncestorLocked(shape)) {
        continue;
      }
      if (util.hitTestLineSegment(
        shape,
        this.editor.getPointInShapeSpace(shape, previousPagePoint),
        this.editor.getPointInShapeSpace(shape, currentPagePoint)
      )) {
        const outermostShape = this.editor.getOutermostSelectableShape(shape);
        const pageMask = this.editor.getPageMaskById(outermostShape.id);
        if (pageMask) {
          const intersection = (0, import_primitives.intersectLineSegmentPolyline)(
            previousPagePoint,
            currentPagePoint,
            pageMask
          );
          if (intersection !== null) {
            const isInMask = (0, import_primitives.pointInPolygon)(currentPagePoint, pageMask);
            if (!isInMask)
              continue;
          }
        }
        this.newlySelectedIds.add(outermostShape.id);
      }
    }
    this.editor.setSelectedIds(
      [.../* @__PURE__ */ new Set([...this.newlySelectedIds, ...this.initialSelectedIds])],
      true
    );
  }
  onCancel = () => {
    this.cancel();
  };
  onComplete = () => {
    this.complete();
  };
  complete() {
    this.parent.transition("idle", {});
  }
  cancel() {
    this.editor.setSelectedIds([...this.initialSelectedIds], true);
    this.parent.transition("idle", {});
  }
}
