// @tldraw/editor/dist-esm/lib/editor/managers/DragAndDropManager.mjs
import { compact } from "@tldraw/utils";
const LAG_DURATION = 100;
class DragAndDropManager {
  constructor(editor) {
    this.editor = editor;
    editor.disposables.add(this.dispose);
  }
  prevDroppingShapeId = null;
  currDroppingShapeId = null;
  droppingNodeTimer = null;
  updateDroppingNode(movingShapes, cb) {
    if (this.droppingNodeTimer === null) {
      const { currentPagePoint } = this.editor.inputs;
      this.currDroppingShapeId = this.editor.getDroppingShape(currentPagePoint, movingShapes)?.id ?? null;
      this.setDragTimer(movingShapes, LAG_DURATION * 10, cb);
    } else if (this.editor.inputs.pointerVelocity.len() > 0.5) {
      clearInterval(this.droppingNodeTimer);
      this.setDragTimer(movingShapes, LAG_DURATION, cb);
    }
  }
  setDragTimer(movingShapes, duration, cb) {
    this.droppingNodeTimer = setTimeout(() => {
      this.editor.batch(() => {
        this.handleDrag(movingShapes, cb);
      });
      this.droppingNodeTimer = null;
    }, duration);
  }
  handleDrag(movingShapes, cb) {
    const { currentPagePoint } = this.editor.inputs;
    movingShapes = compact(movingShapes.map((shape) => this.editor.getShapeById(shape.id)));
    const currDroppingShapeId = this.editor.getDroppingShape(currentPagePoint, movingShapes)?.id ?? null;
    if (currDroppingShapeId !== this.currDroppingShapeId) {
      this.prevDroppingShapeId = this.currDroppingShapeId;
      this.currDroppingShapeId = currDroppingShapeId;
    }
    const { prevDroppingShapeId } = this;
    if (currDroppingShapeId === prevDroppingShapeId) {
      return;
    }
    const prevDroppingShape = prevDroppingShapeId && this.editor.getShapeById(prevDroppingShapeId);
    const nextDroppingShape = currDroppingShapeId && this.editor.getShapeById(currDroppingShapeId);
    this.prevDroppingShapeId = this.currDroppingShapeId;
    if (prevDroppingShape) {
      this.editor.getShapeUtil(prevDroppingShape).onDragShapesOut?.(prevDroppingShape, movingShapes);
    }
    if (nextDroppingShape) {
      const res = this.editor.getShapeUtil(nextDroppingShape).onDragShapesOver?.(nextDroppingShape, movingShapes);
      if (res && res.shouldHint) {
        this.editor.setHintingIds([nextDroppingShape.id]);
      }
    } else {
      this.editor.setHintingIds([]);
    }
    cb?.();
  }
  dropShapes(shapes) {
    const { currDroppingShapeId } = this;
    this.handleDrag(shapes);
    if (currDroppingShapeId) {
      const shape = this.editor.getShapeById(currDroppingShapeId);
      if (!shape)
        return;
      this.editor.getShapeUtil(shape).onDropShapesOver?.(shape, shapes);
    }
  }
  clear() {
    this.prevDroppingShapeId = null;
    this.currDroppingShapeId = null;
    if (this.droppingNodeTimer !== null) {
      clearInterval(this.droppingNodeTimer);
    }
    this.droppingNodeTimer = null;
    this.editor.setHintingIds([]);
  }
  dispose = () => {
    this.clear();
  };
}
export {
  DragAndDropManager
};
