// @tldraw/editor/dist-cjs/lib/editor/shapes/line/LineShapeUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LineShapeUtil_exports = {};
__export(LineShapeUtil_exports, {
  LineShapeUtil: () => LineShapeUtil,
  getSplineForLineShape: () => getSplineForLineShape
});
module.exports = __toCommonJS(LineShapeUtil_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_indices = require("@tldraw/indices");
var import_primitives = require("@tldraw/primitives");
var import_utils = require("@tldraw/utils");
var import_SVGContainer = require("../../../components/SVGContainer");
var import_WeakMapCache = require("../../../utils/WeakMapCache");
var import_ShapeUtil = require("../ShapeUtil");
var import_ShapeFill = require("../shared/ShapeFill");
var import_default_shape_constants = require("../shared/default-shape-constants");
var import_getPerfectDashProps = require("../shared/getPerfectDashProps");
var import_useForceSolid = require("../shared/useForceSolid");
var import_getLinePath = require("./components/getLinePath");
var import_getLineSvg = require("./components/getLineSvg");
const splinesCache = new import_WeakMapCache.WeakMapCache();
const handlesCache = new import_WeakMapCache.WeakMapCache();
class LineShapeUtil extends import_ShapeUtil.ShapeUtil {
  static type = "line";
  hideResizeHandles = () => true;
  hideRotateHandle = () => true;
  hideSelectionBoundsBg = () => true;
  hideSelectionBoundsFg = () => true;
  isClosed = () => false;
  defaultProps() {
    return {
      dash: "draw",
      size: "m",
      color: "black",
      spline: "line",
      handles: {
        start: {
          id: "start",
          type: "vertex",
          canBind: false,
          index: "a1",
          x: 0,
          y: 0
        },
        end: {
          id: "end",
          type: "vertex",
          canBind: false,
          index: "a2",
          x: 0,
          y: 0
        }
      }
    };
  }
  getBounds(shape) {
    const spline = getSplineForLineShape(shape);
    return spline.bounds;
  }
  getCenter(shape) {
    return this.bounds(shape).center;
  }
  getHandles(shape) {
    return handlesCache.get(shape.props, () => {
      const handles = shape.props.handles;
      const spline = getSplineForLineShape(shape);
      const sortedHandles = Object.values(handles).sort(import_indices.sortByIndex);
      const results = sortedHandles.slice();
      for (let i = 0; i < spline.segments.length; i++) {
        const segment = spline.segments[i];
        const point = segment.midPoint;
        const index = (0, import_indices.getIndexBetween)(sortedHandles[i].index, sortedHandles[i + 1].index);
        results.push({
          id: `mid-${i}`,
          type: "create",
          index,
          x: point.x,
          y: point.y
        });
      }
      return results.sort(import_indices.sortByIndex);
    });
  }
  getOutline(shape) {
    return (0, import_getLinePath.getLinePoints)(getSplineForLineShape(shape));
  }
  getOutlineSegments(shape) {
    const spline = getSplineForLineShape(shape);
    return shape.props.spline === "cubic" ? spline.segments.map((s) => s.lut) : spline.segments.map((s) => [s.getPoint(0), s.getPoint(1)]);
  }
  //   Events
  onResize = (shape, info) => {
    const { scaleX, scaleY } = info;
    const handles = (0, import_utils.deepCopy)(shape.props.handles);
    Object.values(shape.props.handles).forEach(({ id, x, y }) => {
      handles[id].x = x * scaleX;
      handles[id].y = y * scaleY;
    });
    return {
      props: {
        handles
      }
    };
  };
  onHandleChange = (shape, { handle }) => {
    const next = (0, import_utils.deepCopy)(shape);
    switch (handle.id) {
      case "start":
      case "end": {
        next.props.handles[handle.id] = {
          ...next.props.handles[handle.id],
          x: handle.x,
          y: handle.y
        };
        break;
      }
      default: {
        const id = "handle:" + handle.index;
        const existing = shape.props.handles[id];
        if (existing) {
          next.props.handles[id] = {
            ...existing,
            x: handle.x,
            y: handle.y
          };
        } else {
          next.props.handles[id] = {
            id,
            type: "vertex",
            canBind: false,
            index: handle.index,
            x: handle.x,
            y: handle.y
          };
        }
        break;
      }
    }
    return next;
  };
  hitTestPoint(shape, point) {
    const zoomLevel = this.editor.zoomLevel;
    const offsetDist = import_default_shape_constants.STROKE_SIZES[shape.props.size] / zoomLevel;
    return (0, import_primitives.pointNearToPolyline)(point, this.outline(shape), offsetDist);
  }
  hitTestLineSegment(shape, A, B) {
    return (0, import_primitives.intersectLineSegmentPolyline)(A, B, this.outline(shape)) !== null;
  }
  component(shape) {
    const forceSolid = (0, import_useForceSolid.useForceSolid)();
    const spline = getSplineForLineShape(shape);
    const strokeWidth = import_default_shape_constants.STROKE_SIZES[shape.props.size];
    const { dash, color } = shape.props;
    if (shape.props.spline === "line") {
      if (dash === "solid") {
        const outline = spline.points;
        const pathData = "M" + outline[0] + "L" + outline.slice(1);
        return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_SVGContainer.SVGContainer, { id: shape.id, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ShapeFill.ShapeFill, { d: pathData, fill: "none", color }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "path",
            {
              d: pathData,
              stroke: `var(--palette-${color})`,
              strokeWidth,
              fill: "none"
            }
          )
        ] });
      }
      if (dash === "dashed" || dash === "dotted") {
        const outline = spline.points;
        const pathData = "M" + outline[0] + "L" + outline.slice(1);
        return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_SVGContainer.SVGContainer, { id: shape.id, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ShapeFill.ShapeFill, { d: pathData, fill: "none", color }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("g", { stroke: `var(--palette-${color})`, strokeWidth, children: spline.segments.map((segment, i) => {
            const { strokeDasharray, strokeDashoffset } = (0, import_getPerfectDashProps.getPerfectDashProps)(
              segment.length,
              strokeWidth,
              {
                style: dash,
                start: i > 0 ? "outset" : "none",
                end: i < spline.segments.length - 1 ? "outset" : "none"
              }
            );
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "path",
              {
                strokeDasharray,
                strokeDashoffset,
                d: segment.path,
                fill: "none"
              },
              i
            );
          }) })
        ] });
      }
      if (dash === "draw") {
        const outline = spline.points;
        const [innerPathData, outerPathData] = (0, import_primitives.getDrawLinePathData)(shape.id, outline, strokeWidth);
        return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_SVGContainer.SVGContainer, { id: shape.id, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ShapeFill.ShapeFill, { d: innerPathData, fill: "none", color }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "path",
            {
              d: outerPathData,
              stroke: `var(--palette-${color})`,
              strokeWidth,
              fill: "none"
            }
          )
        ] });
      }
    }
    if (shape.props.spline === "cubic") {
      const splinePath = spline.path;
      if (dash === "solid" || dash === "draw" && forceSolid) {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_SVGContainer.SVGContainer, { id: shape.id, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ShapeFill.ShapeFill, { d: splinePath, fill: "none", color }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "path",
            {
              strokeWidth,
              stroke: `var(--palette-${color})`,
              fill: "none",
              d: splinePath
            }
          )
        ] });
      }
      if (dash === "dashed" || dash === "dotted") {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_SVGContainer.SVGContainer, { id: shape.id, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ShapeFill.ShapeFill, { d: splinePath, fill: "none", color }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("g", { stroke: `var(--palette-${color})`, strokeWidth, children: spline.segments.map((segment, i) => {
            const { strokeDasharray, strokeDashoffset } = (0, import_getPerfectDashProps.getPerfectDashProps)(
              segment.length,
              strokeWidth,
              {
                style: dash,
                start: i > 0 ? "outset" : "none",
                end: i < spline.segments.length - 1 ? "outset" : "none"
              }
            );
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              "path",
              {
                strokeDasharray,
                strokeDashoffset,
                d: segment.path,
                fill: "none"
              },
              i
            );
          }) })
        ] });
      }
      if (dash === "draw") {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_SVGContainer.SVGContainer, { id: shape.id, children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_ShapeFill.ShapeFill, { d: splinePath, fill: "none", color }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "path",
            {
              d: (0, import_getLinePath.getLineDrawPath)(shape, spline, strokeWidth),
              strokeWidth: 1,
              stroke: `var(--palette-${color})`,
              fill: `var(--palette-${color})`
            }
          )
        ] });
      }
    }
  }
  indicator(shape) {
    const strokeWidth = import_default_shape_constants.STROKE_SIZES[shape.props.size];
    const spline = getSplineForLineShape(shape);
    const { dash } = shape.props;
    let path;
    if (shape.props.spline === "line") {
      const outline = spline.points;
      if (dash === "solid" || dash === "dotted" || dash === "dashed") {
        path = "M" + outline[0] + "L" + outline.slice(1);
      } else {
        const [innerPathData] = (0, import_primitives.getDrawLinePathData)(shape.id, outline, strokeWidth);
        path = innerPathData;
      }
    } else {
      path = (0, import_getLinePath.getLineIndicatorPath)(shape, spline, strokeWidth);
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("path", { d: path });
  }
  toSvg(shape, _font, colors) {
    const { color: _color, size } = shape.props;
    const color = colors.fill[_color];
    const spline = getSplineForLineShape(shape);
    return (0, import_getLineSvg.getLineSvg)(shape, spline, color, import_default_shape_constants.STROKE_SIZES[size]);
  }
}
function getSplineForLineShape(shape) {
  return splinesCache.get(shape.props, () => {
    const { spline, handles } = shape.props;
    const handlePoints = Object.values(handles).sort(import_indices.sortByIndex).map(import_primitives.Vec2d.From);
    switch (spline) {
      case "cubic": {
        return new import_primitives.CubicSpline2d(handlePoints, handlePoints.length === 2 ? 2 : 1.2, 20);
      }
      case "line": {
        return new import_primitives.Polyline2d(handlePoints);
      }
    }
  });
}
