// @tldraw/editor/dist-esm/lib/editor/shapes/highlight/HighlightShapeUtil.mjs
import { jsx } from "react/jsx-runtime";
import { Box2d, getStrokePoints, linesIntersect, Vec2d } from "@tldraw/primitives";
import { last, rng } from "@tldraw/utils";
import { SVGContainer } from "../../../components/SVGContainer.mjs";
import { getSvgPathFromStrokePoints } from "../../../utils/svg.mjs";
import { getHighlightFreehandSettings, getPointsFromSegments } from "../draw/getPath.mjs";
import { ShapeUtil } from "../ShapeUtil.mjs";
import { FONT_SIZES } from "../shared/default-shape-constants.mjs";
import { useForceSolid } from "../shared/useForceSolid.mjs";
const OVERLAY_OPACITY = 0.35;
const UNDERLAY_OPACITY = 0.82;
class HighlightShapeUtil extends ShapeUtil {
  static type = "highlight";
  hideResizeHandles = (shape) => getIsDot(shape);
  hideRotateHandle = (shape) => getIsDot(shape);
  hideSelectionBoundsBg = (shape) => getIsDot(shape);
  hideSelectionBoundsFg = (shape) => getIsDot(shape);
  defaultProps() {
    return {
      segments: [],
      color: "black",
      size: "m",
      isComplete: false,
      isPen: false
    };
  }
  getBounds(shape) {
    return Box2d.FromPoints(this.outline(shape));
  }
  getOutline(shape) {
    return getPointsFromSegments(shape.props.segments);
  }
  getCenter(shape) {
    return this.bounds(shape).center;
  }
  hitTestPoint(shape, point) {
    const outline = this.outline(shape);
    const zoomLevel = this.editor.zoomLevel;
    const offsetDist = getStrokeWidth(shape) / zoomLevel;
    if (shape.props.segments.length === 1 && shape.props.segments[0].points.length < 4) {
      if (shape.props.segments[0].points.some((pt) => Vec2d.Dist(point, pt) < offsetDist * 1.5)) {
        return true;
      }
    }
    if (this.bounds(shape).containsPoint(point)) {
      for (let i = 0; i < outline.length; i++) {
        const C = outline[i];
        const D = outline[(i + 1) % outline.length];
        if (Vec2d.DistanceToLineSegment(C, D, point) < offsetDist)
          return true;
      }
    }
    return false;
  }
  hitTestLineSegment(shape, A, B) {
    const outline = this.outline(shape);
    if (shape.props.segments.length === 1 && shape.props.segments[0].points.length < 4) {
      const zoomLevel = this.editor.zoomLevel;
      const offsetDist = getStrokeWidth(shape) / zoomLevel;
      if (shape.props.segments[0].points.some(
        (pt) => Vec2d.DistanceToLineSegment(A, B, pt) < offsetDist * 1.5
      )) {
        return true;
      }
    }
    for (let i = 0; i < outline.length - 1; i++) {
      const C = outline[i];
      const D = outline[i + 1];
      if (linesIntersect(A, B, C, D))
        return true;
    }
    return false;
  }
  component(shape) {
    return /* @__PURE__ */ jsx(
      HighlightRenderer,
      {
        strokeWidth: getStrokeWidth(shape),
        shape,
        opacity: OVERLAY_OPACITY
      }
    );
  }
  renderBackground(shape) {
    return /* @__PURE__ */ jsx(
      HighlightRenderer,
      {
        strokeWidth: getStrokeWidth(shape),
        shape,
        opacity: UNDERLAY_OPACITY
      }
    );
  }
  indicator(shape) {
    const forceSolid = useForceSolid();
    const strokeWidth = getStrokeWidth(shape);
    const allPointsFromSegments = getPointsFromSegments(shape.props.segments);
    let sw = strokeWidth;
    if (!forceSolid && !shape.props.isPen && allPointsFromSegments.length === 1) {
      sw += rng(shape.id)() * (strokeWidth / 6);
    }
    const showAsComplete = shape.props.isComplete || last(shape.props.segments)?.type === "straight";
    const options = getHighlightFreehandSettings({
      strokeWidth,
      showAsComplete,
      isPen: shape.props.isPen
    });
    const strokePoints = getStrokePoints(allPointsFromSegments, options);
    let strokePath;
    if (strokePoints.length < 2) {
      strokePath = getIndicatorDot(allPointsFromSegments[0], sw);
    } else {
      strokePath = getSvgPathFromStrokePoints(strokePoints, false);
    }
    return /* @__PURE__ */ jsx("path", { d: strokePath });
  }
  expandSelectionOutlinePx(shape) {
    return getStrokeWidth(shape) / 2;
  }
  toSvg(shape, _font, colors) {
    return highlighterToSvg(getStrokeWidth(shape), shape, OVERLAY_OPACITY, colors);
  }
  toBackgroundSvg(shape, font, colors) {
    return highlighterToSvg(getStrokeWidth(shape), shape, UNDERLAY_OPACITY, colors);
  }
  onResize = (shape, info) => {
    const { scaleX, scaleY } = info;
    const newSegments = [];
    for (const segment of shape.props.segments) {
      newSegments.push({
        ...segment,
        points: segment.points.map(({ x, y, z }) => {
          return {
            x: scaleX * x,
            y: scaleY * y,
            z
          };
        })
      });
    }
    return {
      props: {
        segments: newSegments
      }
    };
  };
}
function getShapeDot(point) {
  const r = 0.1;
  return `M ${point.x} ${point.y} m -${r}, 0 a ${r},${r} 0 1,0 ${r * 2},0 a ${r},${r} 0 1,0 -${r * 2},0`;
}
function getIndicatorDot(point, sw) {
  const r = sw / 2;
  return `M ${point.x} ${point.y} m -${r}, 0 a ${r},${r} 0 1,0 ${r * 2},0 a ${r},${r} 0 1,0 -${r * 2},0`;
}
function getHighlightSvgPath(shape, strokeWidth, forceSolid) {
  const allPointsFromSegments = getPointsFromSegments(shape.props.segments);
  const showAsComplete = shape.props.isComplete || last(shape.props.segments)?.type === "straight";
  let sw = strokeWidth;
  if (!forceSolid && !shape.props.isPen && allPointsFromSegments.length === 1) {
    sw += rng(shape.id)() * (strokeWidth / 6);
  }
  const options = getHighlightFreehandSettings({
    strokeWidth: sw,
    showAsComplete,
    isPen: shape.props.isPen
  });
  const strokePoints = getStrokePoints(allPointsFromSegments, options);
  const solidStrokePath = strokePoints.length > 1 ? getSvgPathFromStrokePoints(strokePoints, false) : getShapeDot(allPointsFromSegments[0]);
  return { solidStrokePath, sw };
}
function HighlightRenderer({
  strokeWidth,
  shape,
  opacity
}) {
  const forceSolid = useForceSolid();
  const { solidStrokePath, sw } = getHighlightSvgPath(shape, strokeWidth, forceSolid);
  return /* @__PURE__ */ jsx(SVGContainer, { id: shape.id, style: { opacity }, children: /* @__PURE__ */ jsx(
    "path",
    {
      d: solidStrokePath,
      strokeLinecap: "round",
      fill: "none",
      pointerEvents: "all",
      stroke: `var(--palette-${shape.props.color}-highlight)`,
      strokeWidth: sw
    }
  ) });
}
function highlighterToSvg(strokeWidth, shape, opacity, colors) {
  const { solidStrokePath, sw } = getHighlightSvgPath(shape, strokeWidth, false);
  const path = document.createElementNS("http://www.w3.org/2000/svg", "path");
  path.setAttribute("d", solidStrokePath);
  path.setAttribute("fill", "none");
  path.setAttribute("stroke", colors.highlight[shape.props.color]);
  path.setAttribute("stroke-width", `${sw}`);
  path.setAttribute("opacity", `${opacity}`);
  return path;
}
function getStrokeWidth(shape) {
  return FONT_SIZES[shape.props.size] * 1.12;
}
function getIsDot(shape) {
  return shape.props.segments.length === 1 && shape.props.segments[0].points.length < 2;
}
export {
  HighlightShapeUtil
};
