// @tldraw/editor/dist-esm/lib/editor/tools/SelectTool/children/PointingCropHandle.mjs
import { StateNode } from "../../StateNode.mjs";
import { CursorTypeMap } from "./PointingResizeHandle.mjs";
class PointingCropHandle extends StateNode {
  static id = "pointing_crop_handle";
  info = {};
  updateCursor(shape) {
    const cursorType = CursorTypeMap[this.info.handle];
    this.editor.setCursor({
      type: cursorType,
      rotation: shape.rotation
    });
  }
  onEnter = (info) => {
    this.info = info;
    const selectedShape = this.editor.selectedShapes[0];
    if (!selectedShape)
      return;
    this.updateCursor(selectedShape);
    this.editor.setCroppingId(selectedShape.id);
  };
  onPointerMove = () => {
    const isDragging = this.editor.inputs.isDragging;
    if (isDragging) {
      this.parent.transition("cropping", {
        ...this.info,
        onInteractionEnd: this.info.onInteractionEnd
      });
    }
  };
  onPointerUp = () => {
    if (this.info.onInteractionEnd) {
      this.editor.setSelectedTool(this.info.onInteractionEnd, this.info);
    } else {
      this.editor.setCroppingId(null);
      this.parent.transition("idle", {});
    }
  };
  onCancel = () => {
    this.cancel();
  };
  onComplete = () => {
    this.cancel();
  };
  onInterrupt = () => {
    this.cancel();
  };
  cancel() {
    if (this.info.onInteractionEnd) {
      this.editor.setSelectedTool(this.info.onInteractionEnd, this.info);
    } else {
      this.editor.setCroppingId(null);
      this.parent.transition("idle", {});
    }
  }
}
export {
  PointingCropHandle
};
