// @tldraw/editor/dist-cjs/lib/editor/shapes/line/toolStates/Pointing.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Pointing_exports = {};
__export(Pointing_exports, {
  Pointing: () => Pointing
});
module.exports = __toCommonJS(Pointing_exports);
var import_indices = require("@tldraw/indices");
var import_primitives = require("@tldraw/primitives");
var import_tlschema = require("@tldraw/tlschema");
var import_utils = require("@tldraw/utils");
var import_StateNode = require("../../../tools/StateNode");
class Pointing extends import_StateNode.StateNode {
  static id = "pointing";
  shape = {};
  markPointId = "";
  onEnter = (info) => {
    const { inputs } = this.editor;
    const { currentPagePoint } = inputs;
    this.markPointId = this.editor.mark("creating");
    let shapeExists = false;
    if (info.shapeId) {
      const shape = this.editor.getShapeById(info.shapeId);
      if (shape) {
        shapeExists = true;
        this.shape = shape;
      }
    }
    if (inputs.shiftKey && shapeExists) {
      const handles = this.editor.getShapeUtil(this.shape).handles(this.shape);
      const vertexHandles = handles.filter((h) => h.type === "vertex").sort(import_indices.sortByIndex);
      const endHandle = vertexHandles[vertexHandles.length - 1];
      const shapePagePoint = import_primitives.Matrix2d.applyToPoint(
        this.editor.getParentTransform(this.shape),
        new import_primitives.Vec2d(this.shape.x, this.shape.y)
      );
      let nextEndHandleIndex, nextEndHandleId, nextEndHandle;
      if (vertexHandles.length === 2 && vertexHandles[1].x === 1 && vertexHandles[1].y === 1) {
        nextEndHandleIndex = vertexHandles[1].index;
        nextEndHandleId = vertexHandles[1].id;
        nextEndHandle = {
          ...vertexHandles[1],
          x: currentPagePoint.x - shapePagePoint.x,
          y: currentPagePoint.y - shapePagePoint.y
        };
      } else {
        nextEndHandleIndex = (0, import_indices.getIndexAbove)(endHandle.index);
        nextEndHandleId = "handle:" + nextEndHandleIndex;
        nextEndHandle = {
          x: currentPagePoint.x - shapePagePoint.x,
          y: currentPagePoint.y - shapePagePoint.y,
          index: nextEndHandleIndex,
          canBind: false,
          type: "vertex",
          id: nextEndHandleId
        };
      }
      const nextHandles = (0, import_utils.structuredClone)(this.shape.props.handles);
      nextHandles[nextEndHandle.id] = nextEndHandle;
      this.editor.updateShapes([
        {
          id: this.shape.id,
          type: this.shape.type,
          props: {
            handles: nextHandles
          }
        }
      ]);
    } else {
      const id = (0, import_tlschema.createShapeId)();
      this.editor.createShapes([
        {
          id,
          type: "line",
          x: currentPagePoint.x,
          y: currentPagePoint.y
        }
      ]);
      this.editor.select(id);
      this.shape = this.editor.getShapeById(id);
    }
  };
  onPointerMove = () => {
    if (!this.shape)
      return;
    if (this.editor.inputs.isDragging) {
      const util = this.editor.getShapeUtil(this.shape);
      const handles = util.handles?.(this.shape);
      if (!handles) {
        this.editor.bailToMark("creating");
        throw Error("No handles found");
      }
      this.editor.setSelectedTool("select.dragging_handle", {
        shape: this.shape,
        isCreating: true,
        handle: (0, import_utils.last)(handles),
        onInteractionEnd: "line"
      });
    }
  };
  onPointerUp = () => {
    this.complete();
  };
  onCancel = () => {
    this.cancel();
  };
  onComplete = () => {
    this.complete();
  };
  onInterrupt = () => {
    this.parent.transition("idle", {});
    this.editor.bailToMark("creating");
    this.editor.snaps.clear();
  };
  complete() {
    this.parent.transition("idle", { shapeId: this.shape.id });
    this.editor.snaps.clear();
  }
  cancel() {
    this.editor.bailToMark(this.markPointId);
    this.parent.transition("idle", { shapeId: this.shape.id });
    this.editor.snaps.clear();
  }
}
