// signia-react/dist/cjs/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  track: () => import_track.track,
  useAtom: () => import_useAtom.useAtom,
  useComputed: () => import_useComputed.useComputed,
  useStateTracking: () => import_useStateTracking.useStateTracking,
  useValue: () => import_useValue.useValue,
  wrapJsx: () => import_wrapJsx.wrapJsx
});
module.exports = __toCommonJS(src_exports);
var import_track = require("./track.cjs");
var import_useAtom = require("./useAtom.cjs");
var import_useComputed = require("./useComputed.cjs");
var import_useStateTracking = require("./useStateTracking.cjs");
var import_useValue = require("./useValue.cjs");
var import_wrapJsx = require("./wrapJsx.cjs");
