// @tldraw/editor/dist-cjs/lib/editor/shapes/draw/DrawShapeUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DrawShapeUtil_exports = {};
__export(DrawShapeUtil_exports, {
  DrawShapeUtil: () => DrawShapeUtil
});
module.exports = __toCommonJS(DrawShapeUtil_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_primitives = require("@tldraw/primitives");
var import_utils = require("@tldraw/utils");
var import_SVGContainer = require("../../../components/SVGContainer");
var import_svg = require("../../../utils/svg");
var import_ShapeUtil = require("../ShapeUtil");
var import_default_shape_constants = require("../shared/default-shape-constants");
var import_ShapeFill = require("../shared/ShapeFill");
var import_useForceSolid = require("../shared/useForceSolid");
var import_getPath = require("./getPath");
class DrawShapeUtil extends import_ShapeUtil.ShapeUtil {
  static type = "draw";
  hideResizeHandles = (shape) => getIsDot(shape);
  hideRotateHandle = (shape) => getIsDot(shape);
  hideSelectionBoundsBg = (shape) => getIsDot(shape);
  hideSelectionBoundsFg = (shape) => getIsDot(shape);
  defaultProps() {
    return {
      segments: [],
      color: "black",
      fill: "none",
      dash: "draw",
      size: "m",
      isComplete: false,
      isClosed: false,
      isPen: false
    };
  }
  isClosed = (shape) => shape.props.isClosed;
  getBounds(shape) {
    return import_primitives.Box2d.FromPoints(this.outline(shape));
  }
  getOutline(shape) {
    return (0, import_getPath.getPointsFromSegments)(shape.props.segments);
  }
  getCenter(shape) {
    return this.bounds(shape).center;
  }
  hitTestPoint(shape, point) {
    const outline = this.outline(shape);
    const zoomLevel = this.editor.zoomLevel;
    const offsetDist = import_default_shape_constants.STROKE_SIZES[shape.props.size] / zoomLevel;
    if (shape.props.segments.length === 1 && shape.props.segments[0].points.length < 4) {
      if (shape.props.segments[0].points.some((pt) => import_primitives.Vec2d.Dist(point, pt) < offsetDist * 1.5)) {
        return true;
      }
    }
    if (this.isClosed(shape)) {
      return (0, import_primitives.pointInPolygon)(point, outline);
    }
    if (this.bounds(shape).containsPoint(point)) {
      for (let i = 0; i < outline.length; i++) {
        const C = outline[i];
        const D = outline[(i + 1) % outline.length];
        if (import_primitives.Vec2d.DistanceToLineSegment(C, D, point) < offsetDist)
          return true;
      }
    }
    return false;
  }
  hitTestLineSegment(shape, A, B) {
    const outline = this.outline(shape);
    if (shape.props.segments.length === 1 && shape.props.segments[0].points.length < 4) {
      const zoomLevel = this.editor.zoomLevel;
      const offsetDist = import_default_shape_constants.STROKE_SIZES[shape.props.size] / zoomLevel;
      if (shape.props.segments[0].points.some(
        (pt) => import_primitives.Vec2d.DistanceToLineSegment(A, B, pt) < offsetDist * 1.5
      )) {
        return true;
      }
    }
    if (this.isClosed(shape)) {
      for (let i = 0; i < outline.length; i++) {
        const C = outline[i];
        const D = outline[(i + 1) % outline.length];
        if ((0, import_primitives.linesIntersect)(A, B, C, D))
          return true;
      }
    } else {
      for (let i = 0; i < outline.length - 1; i++) {
        const C = outline[i];
        const D = outline[i + 1];
        if ((0, import_primitives.linesIntersect)(A, B, C, D))
          return true;
      }
    }
    return false;
  }
  component(shape) {
    const forceSolid = (0, import_useForceSolid.useForceSolid)();
    const strokeWidth = import_default_shape_constants.STROKE_SIZES[shape.props.size];
    const allPointsFromSegments = (0, import_getPath.getPointsFromSegments)(shape.props.segments);
    const showAsComplete = shape.props.isComplete || (0, import_utils.last)(shape.props.segments)?.type === "straight";
    let sw = strokeWidth;
    if (!forceSolid && !shape.props.isPen && shape.props.dash === "draw" && allPointsFromSegments.length === 1) {
      sw += (0, import_utils.rng)(shape.id)() * (strokeWidth / 6);
    }
    const options = (0, import_getPath.getFreehandOptions)(shape.props, sw, showAsComplete, forceSolid);
    const strokePoints = (0, import_primitives.getStrokePoints)(allPointsFromSegments, options);
    const solidStrokePath = strokePoints.length > 1 ? (0, import_svg.getSvgPathFromStrokePoints)(strokePoints, shape.props.isClosed) : getDot(allPointsFromSegments[0], sw);
    if (!forceSolid && shape.props.dash === "draw" || strokePoints.length < 2) {
      (0, import_primitives.setStrokePointRadii)(strokePoints, options);
      const strokeOutlinePoints = (0, import_primitives.getStrokeOutlinePoints)(strokePoints, options);
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_SVGContainer.SVGContainer, { id: shape.id, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_ShapeFill.ShapeFill,
          {
            fill: shape.props.isClosed ? shape.props.fill : "none",
            color: shape.props.color,
            d: solidStrokePath
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "path",
          {
            d: (0, import_svg.getSvgPathFromStroke)(strokeOutlinePoints, true),
            strokeLinecap: "round",
            fill: `var(--palette-${shape.props.color})`
          }
        )
      ] });
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_SVGContainer.SVGContainer, { id: shape.id, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_ShapeFill.ShapeFill,
        {
          color: shape.props.color,
          fill: shape.props.isClosed ? shape.props.fill : "none",
          d: solidStrokePath
        }
      ),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "path",
        {
          d: solidStrokePath,
          strokeLinecap: "round",
          fill: "none",
          stroke: `var(--palette-${shape.props.color})`,
          strokeWidth,
          strokeDasharray: (0, import_getPath.getDrawShapeStrokeDashArray)(shape, strokeWidth),
          strokeDashoffset: "0"
        }
      )
    ] });
  }
  indicator(shape) {
    const forceSolid = (0, import_useForceSolid.useForceSolid)();
    const strokeWidth = import_default_shape_constants.STROKE_SIZES[shape.props.size];
    const allPointsFromSegments = (0, import_getPath.getPointsFromSegments)(shape.props.segments);
    let sw = strokeWidth;
    if (!forceSolid && !shape.props.isPen && shape.props.dash === "draw" && allPointsFromSegments.length === 1) {
      sw += (0, import_utils.rng)(shape.id)() * (strokeWidth / 6);
    }
    const showAsComplete = shape.props.isComplete || (0, import_utils.last)(shape.props.segments)?.type === "straight";
    const options = (0, import_getPath.getFreehandOptions)(shape.props, sw, showAsComplete, true);
    const strokePoints = (0, import_primitives.getStrokePoints)(allPointsFromSegments, options);
    const solidStrokePath = strokePoints.length > 1 ? (0, import_svg.getSvgPathFromStrokePoints)(strokePoints, shape.props.isClosed) : getDot(allPointsFromSegments[0], sw);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("path", { d: solidStrokePath });
  }
  toSvg(shape, _font, colors) {
    const { color } = shape.props;
    const strokeWidth = import_default_shape_constants.STROKE_SIZES[shape.props.size];
    const allPointsFromSegments = (0, import_getPath.getPointsFromSegments)(shape.props.segments);
    const showAsComplete = shape.props.isComplete || (0, import_utils.last)(shape.props.segments)?.type === "straight";
    let sw = strokeWidth;
    if (!shape.props.isPen && shape.props.dash === "draw" && allPointsFromSegments.length === 1) {
      sw += (0, import_utils.rng)(shape.id)() * (strokeWidth / 6);
    }
    const options = (0, import_getPath.getFreehandOptions)(shape.props, sw, showAsComplete, false);
    const strokePoints = (0, import_primitives.getStrokePoints)(allPointsFromSegments, options);
    const solidStrokePath = strokePoints.length > 1 ? (0, import_svg.getSvgPathFromStrokePoints)(strokePoints, shape.props.isClosed) : getDot(allPointsFromSegments[0], sw);
    let foregroundPath;
    if (shape.props.dash === "draw" || strokePoints.length < 2) {
      (0, import_primitives.setStrokePointRadii)(strokePoints, options);
      const strokeOutlinePoints = (0, import_primitives.getStrokeOutlinePoints)(strokePoints, options);
      const p = document.createElementNS("http://www.w3.org/2000/svg", "path");
      p.setAttribute("d", (0, import_svg.getSvgPathFromStroke)(strokeOutlinePoints, true));
      p.setAttribute("fill", colors.fill[color]);
      p.setAttribute("stroke-linecap", "round");
      foregroundPath = p;
    } else {
      const p = document.createElementNS("http://www.w3.org/2000/svg", "path");
      p.setAttribute("d", solidStrokePath);
      p.setAttribute("stroke", colors.fill[color]);
      p.setAttribute("fill", "none");
      p.setAttribute("stroke-linecap", "round");
      p.setAttribute("stroke-width", strokeWidth.toString());
      p.setAttribute("stroke-dasharray", (0, import_getPath.getDrawShapeStrokeDashArray)(shape, strokeWidth));
      p.setAttribute("stroke-dashoffset", "0");
      foregroundPath = p;
    }
    const fillPath = (0, import_ShapeFill.getShapeFillSvg)({
      fill: shape.props.isClosed ? shape.props.fill : "none",
      d: solidStrokePath,
      color: shape.props.color,
      colors
    });
    if (fillPath) {
      const g = document.createElementNS("http://www.w3.org/2000/svg", "g");
      g.appendChild(fillPath);
      g.appendChild(foregroundPath);
      return g;
    }
    return foregroundPath;
  }
  onResize = (shape, info) => {
    const { scaleX, scaleY } = info;
    const newSegments = [];
    for (const segment of shape.props.segments) {
      newSegments.push({
        ...segment,
        points: segment.points.map(({ x, y, z }) => {
          return {
            x: (0, import_primitives.toFixed)(scaleX * x),
            y: (0, import_primitives.toFixed)(scaleY * y),
            z
          };
        })
      });
    }
    return {
      props: {
        segments: newSegments
      }
    };
  };
  expandSelectionOutlinePx(shape) {
    const multiplier = shape.props.dash === "draw" ? 1.6 : 1;
    return import_default_shape_constants.STROKE_SIZES[shape.props.size] * multiplier / 2;
  }
}
function getDot(point, sw) {
  const r = (sw + 1) * 0.5;
  return `M ${point.x} ${point.y} m -${r}, 0 a ${r},${r} 0 1,0 ${r * 2},0 a ${r},${r} 0 1,0 -${r * 2},0`;
}
function getIsDot(shape) {
  return shape.props.segments.length === 1 && shape.props.segments[0].points.length < 2;
}
