// @tldraw/editor/dist-cjs/lib/editor/shapes/highlight/HighlightShapeUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HighlightShapeUtil_exports = {};
__export(HighlightShapeUtil_exports, {
  HighlightShapeUtil: () => HighlightShapeUtil
});
module.exports = __toCommonJS(HighlightShapeUtil_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_primitives = require("@tldraw/primitives");
var import_utils = require("@tldraw/utils");
var import_SVGContainer = require("../../../components/SVGContainer");
var import_svg = require("../../../utils/svg");
var import_getPath = require("../draw/getPath");
var import_ShapeUtil = require("../ShapeUtil");
var import_default_shape_constants = require("../shared/default-shape-constants");
var import_useForceSolid = require("../shared/useForceSolid");
const OVERLAY_OPACITY = 0.35;
const UNDERLAY_OPACITY = 0.82;
class HighlightShapeUtil extends import_ShapeUtil.ShapeUtil {
  static type = "highlight";
  hideResizeHandles = (shape) => getIsDot(shape);
  hideRotateHandle = (shape) => getIsDot(shape);
  hideSelectionBoundsBg = (shape) => getIsDot(shape);
  hideSelectionBoundsFg = (shape) => getIsDot(shape);
  defaultProps() {
    return {
      segments: [],
      color: "black",
      size: "m",
      isComplete: false,
      isPen: false
    };
  }
  getBounds(shape) {
    return import_primitives.Box2d.FromPoints(this.outline(shape));
  }
  getOutline(shape) {
    return (0, import_getPath.getPointsFromSegments)(shape.props.segments);
  }
  getCenter(shape) {
    return this.bounds(shape).center;
  }
  hitTestPoint(shape, point) {
    const outline = this.outline(shape);
    const zoomLevel = this.editor.zoomLevel;
    const offsetDist = getStrokeWidth(shape) / zoomLevel;
    if (shape.props.segments.length === 1 && shape.props.segments[0].points.length < 4) {
      if (shape.props.segments[0].points.some((pt) => import_primitives.Vec2d.Dist(point, pt) < offsetDist * 1.5)) {
        return true;
      }
    }
    if (this.bounds(shape).containsPoint(point)) {
      for (let i = 0; i < outline.length; i++) {
        const C = outline[i];
        const D = outline[(i + 1) % outline.length];
        if (import_primitives.Vec2d.DistanceToLineSegment(C, D, point) < offsetDist)
          return true;
      }
    }
    return false;
  }
  hitTestLineSegment(shape, A, B) {
    const outline = this.outline(shape);
    if (shape.props.segments.length === 1 && shape.props.segments[0].points.length < 4) {
      const zoomLevel = this.editor.zoomLevel;
      const offsetDist = getStrokeWidth(shape) / zoomLevel;
      if (shape.props.segments[0].points.some(
        (pt) => import_primitives.Vec2d.DistanceToLineSegment(A, B, pt) < offsetDist * 1.5
      )) {
        return true;
      }
    }
    for (let i = 0; i < outline.length - 1; i++) {
      const C = outline[i];
      const D = outline[i + 1];
      if ((0, import_primitives.linesIntersect)(A, B, C, D))
        return true;
    }
    return false;
  }
  component(shape) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      HighlightRenderer,
      {
        strokeWidth: getStrokeWidth(shape),
        shape,
        opacity: OVERLAY_OPACITY
      }
    );
  }
  renderBackground(shape) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      HighlightRenderer,
      {
        strokeWidth: getStrokeWidth(shape),
        shape,
        opacity: UNDERLAY_OPACITY
      }
    );
  }
  indicator(shape) {
    const forceSolid = (0, import_useForceSolid.useForceSolid)();
    const strokeWidth = getStrokeWidth(shape);
    const allPointsFromSegments = (0, import_getPath.getPointsFromSegments)(shape.props.segments);
    let sw = strokeWidth;
    if (!forceSolid && !shape.props.isPen && allPointsFromSegments.length === 1) {
      sw += (0, import_utils.rng)(shape.id)() * (strokeWidth / 6);
    }
    const showAsComplete = shape.props.isComplete || (0, import_utils.last)(shape.props.segments)?.type === "straight";
    const options = (0, import_getPath.getHighlightFreehandSettings)({
      strokeWidth,
      showAsComplete,
      isPen: shape.props.isPen
    });
    const strokePoints = (0, import_primitives.getStrokePoints)(allPointsFromSegments, options);
    let strokePath;
    if (strokePoints.length < 2) {
      strokePath = getIndicatorDot(allPointsFromSegments[0], sw);
    } else {
      strokePath = (0, import_svg.getSvgPathFromStrokePoints)(strokePoints, false);
    }
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("path", { d: strokePath });
  }
  expandSelectionOutlinePx(shape) {
    return getStrokeWidth(shape) / 2;
  }
  toSvg(shape, _font, colors) {
    return highlighterToSvg(getStrokeWidth(shape), shape, OVERLAY_OPACITY, colors);
  }
  toBackgroundSvg(shape, font, colors) {
    return highlighterToSvg(getStrokeWidth(shape), shape, UNDERLAY_OPACITY, colors);
  }
  onResize = (shape, info) => {
    const { scaleX, scaleY } = info;
    const newSegments = [];
    for (const segment of shape.props.segments) {
      newSegments.push({
        ...segment,
        points: segment.points.map(({ x, y, z }) => {
          return {
            x: scaleX * x,
            y: scaleY * y,
            z
          };
        })
      });
    }
    return {
      props: {
        segments: newSegments
      }
    };
  };
}
function getShapeDot(point) {
  const r = 0.1;
  return `M ${point.x} ${point.y} m -${r}, 0 a ${r},${r} 0 1,0 ${r * 2},0 a ${r},${r} 0 1,0 -${r * 2},0`;
}
function getIndicatorDot(point, sw) {
  const r = sw / 2;
  return `M ${point.x} ${point.y} m -${r}, 0 a ${r},${r} 0 1,0 ${r * 2},0 a ${r},${r} 0 1,0 -${r * 2},0`;
}
function getHighlightSvgPath(shape, strokeWidth, forceSolid) {
  const allPointsFromSegments = (0, import_getPath.getPointsFromSegments)(shape.props.segments);
  const showAsComplete = shape.props.isComplete || (0, import_utils.last)(shape.props.segments)?.type === "straight";
  let sw = strokeWidth;
  if (!forceSolid && !shape.props.isPen && allPointsFromSegments.length === 1) {
    sw += (0, import_utils.rng)(shape.id)() * (strokeWidth / 6);
  }
  const options = (0, import_getPath.getHighlightFreehandSettings)({
    strokeWidth: sw,
    showAsComplete,
    isPen: shape.props.isPen
  });
  const strokePoints = (0, import_primitives.getStrokePoints)(allPointsFromSegments, options);
  const solidStrokePath = strokePoints.length > 1 ? (0, import_svg.getSvgPathFromStrokePoints)(strokePoints, false) : getShapeDot(allPointsFromSegments[0]);
  return { solidStrokePath, sw };
}
function HighlightRenderer({
  strokeWidth,
  shape,
  opacity
}) {
  const forceSolid = (0, import_useForceSolid.useForceSolid)();
  const { solidStrokePath, sw } = getHighlightSvgPath(shape, strokeWidth, forceSolid);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SVGContainer.SVGContainer, { id: shape.id, style: { opacity }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "path",
    {
      d: solidStrokePath,
      strokeLinecap: "round",
      fill: "none",
      pointerEvents: "all",
      stroke: `var(--palette-${shape.props.color}-highlight)`,
      strokeWidth: sw
    }
  ) });
}
function highlighterToSvg(strokeWidth, shape, opacity, colors) {
  const { solidStrokePath, sw } = getHighlightSvgPath(shape, strokeWidth, false);
  const path = document.createElementNS("http://www.w3.org/2000/svg", "path");
  path.setAttribute("d", solidStrokePath);
  path.setAttribute("fill", "none");
  path.setAttribute("stroke", colors.highlight[shape.props.color]);
  path.setAttribute("stroke-width", `${sw}`);
  path.setAttribute("opacity", `${opacity}`);
  return path;
}
function getStrokeWidth(shape) {
  return import_default_shape_constants.FONT_SIZES[shape.props.size] * 1.12;
}
function getIsDot(shape) {
  return shape.props.segments.length === 1 && shape.props.segments[0].points.length < 2;
}
