// @tldraw/editor/dist-cjs/lib/editor/shapes/geo/GeoShapeUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GeoShapeUtil_exports = {};
__export(GeoShapeUtil_exports, {
  GeoShapeUtil: () => GeoShapeUtil
});
module.exports = __toCommonJS(GeoShapeUtil_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_primitives = require("@tldraw/primitives");
var import_SVGContainer = require("../../../components/SVGContainer");
var import_BaseBoxShapeUtil = require("../BaseBoxShapeUtil");
var import_default_shape_constants = require("../shared/default-shape-constants");
var import_getTextLabelSvgElement = require("../shared/getTextLabelSvgElement");
var import_HyperlinkButton = require("../shared/HyperlinkButton");
var import_TextLabel = require("../shared/TextLabel");
var import_useForceSolid = require("../shared/useForceSolid");
var import_DashStyleEllipse = require("./components/DashStyleEllipse");
var import_DashStyleOval = require("./components/DashStyleOval");
var import_DashStylePolygon = require("./components/DashStylePolygon");
var import_DrawStyleEllipse = require("./components/DrawStyleEllipse");
var import_DrawStylePolygon = require("./components/DrawStylePolygon");
var import_SolidStyleEllipse = require("./components/SolidStyleEllipse");
var import_SolidStyleOval = require("./components/SolidStyleOval");
var import_SolidStylePolygon = require("./components/SolidStylePolygon");
const LABEL_PADDING = 16;
const MIN_SIZE_WITH_LABEL = 17 * 3;
class GeoShapeUtil extends import_BaseBoxShapeUtil.BaseBoxShapeUtil {
  static type = "geo";
  canEdit = () => true;
  defaultProps() {
    return {
      w: 100,
      h: 100,
      geo: "rectangle",
      color: "black",
      labelColor: "black",
      fill: "none",
      dash: "draw",
      size: "m",
      font: "draw",
      text: "",
      align: "middle",
      verticalAlign: "middle",
      growY: 0,
      url: ""
    };
  }
  hitTestLineSegment(shape, A, B) {
    const outline = this.outline(shape);
    for (let i = 0; i < outline.length; i++) {
      const C = outline[i];
      const D = outline[(i + 1) % outline.length];
      if ((0, import_primitives.linesIntersect)(A, B, C, D))
        return true;
    }
    const lines = getLines(shape.props, 0);
    if (lines !== void 0) {
      for (const [C, D] of lines) {
        if ((0, import_primitives.linesIntersect)(A, B, C, D))
          return true;
      }
    }
    return false;
  }
  hitTestPoint(shape, point) {
    const outline = this.outline(shape);
    if (shape.props.fill === "none") {
      const zoomLevel = this.editor.zoomLevel;
      const offsetDist = import_default_shape_constants.STROKE_SIZES[shape.props.size] / zoomLevel;
      for (let i = 0; i < outline.length; i++) {
        const C = outline[i];
        const D = outline[(i + 1) % outline.length];
        if (import_primitives.Vec2d.DistanceToLineSegment(C, D, point) < offsetDist)
          return true;
      }
      const lines = getLines(shape.props, 1);
      if (lines !== void 0) {
        for (const [C, D] of lines) {
          if (import_primitives.Vec2d.DistanceToLineSegment(C, D, point) < offsetDist)
            return true;
        }
      }
      return false;
    }
    return (0, import_primitives.pointInPolygon)(point, outline);
  }
  getBounds(shape) {
    return new import_primitives.Box2d(0, 0, shape.props.w, shape.props.h + shape.props.growY);
  }
  getCenter(shape) {
    return new import_primitives.Vec2d(shape.props.w / 2, (shape.props.h + shape.props.growY) / 2);
  }
  getOutline(shape) {
    const w = Math.max(1, shape.props.w);
    const h = Math.max(1, shape.props.h + shape.props.growY);
    const cx = w / 2;
    const cy = h / 2;
    switch (shape.props.geo) {
      case "triangle": {
        return [new import_primitives.Vec2d(cx, 0), new import_primitives.Vec2d(w, h), new import_primitives.Vec2d(0, h)];
      }
      case "diamond": {
        return [new import_primitives.Vec2d(cx, 0), new import_primitives.Vec2d(w, cy), new import_primitives.Vec2d(cx, h), new import_primitives.Vec2d(0, cy)];
      }
      case "pentagon": {
        return (0, import_primitives.getPolygonVertices)(w, h, 5);
      }
      case "hexagon": {
        return (0, import_primitives.getPolygonVertices)(w, h, 6);
      }
      case "octagon": {
        return (0, import_primitives.getPolygonVertices)(w, h, 8);
      }
      case "ellipse": {
        const q = Math.pow(cx - cy, 2) / Math.pow(cx + cy, 2);
        const p = import_primitives.PI * (cx + cy) * (1 + 3 * q / (10 + Math.sqrt(4 - 3 * q)));
        let len = Math.max(4, Math.ceil(p / 10));
        len = Math.ceil(len / 4) * 4;
        const step = import_primitives.PI2 / len;
        const a = Math.cos(step);
        const b = Math.sin(step);
        let sin = 0;
        let cos = 1;
        let ts = 0;
        let tc = 1;
        const points = Array(len);
        for (let i = 0; i < len; i++) {
          points[i] = new import_primitives.Vec2d(cx + cx * cos, cy + cy * sin);
          ts = b * cos + a * sin;
          tc = a * cos - b * sin;
          sin = ts;
          cos = tc;
        }
        return points;
      }
      case "oval": {
        const len = 10;
        const points = Array(len * 2);
        if (h > w) {
          for (let i = 0; i < len; i++) {
            const t1 = -import_primitives.PI + import_primitives.PI * i / (len - 2);
            const t2 = import_primitives.PI * i / (len - 2);
            points[i] = new import_primitives.Vec2d(cx + cx * Math.cos(t1), cx + cx * Math.sin(t1));
            points[i + len] = new import_primitives.Vec2d(cx + cx * Math.cos(t2), h - cx + cx * Math.sin(t2));
          }
        } else {
          for (let i = 0; i < len; i++) {
            const t1 = -import_primitives.TAU + import_primitives.PI * i / (len - 2);
            const t2 = import_primitives.TAU + import_primitives.PI * -i / (len - 2);
            points[i] = new import_primitives.Vec2d(w - cy + cy * Math.cos(t1), h - cy + cy * Math.sin(t1));
            points[i + len] = new import_primitives.Vec2d(cy - cy * Math.cos(t2), h - cy + cy * Math.sin(t2));
          }
        }
        return points;
      }
      case "star": {
        const sides = 5;
        const step = import_primitives.PI2 / sides / 2;
        const rightMostIndex = Math.floor(sides / 4) * 2;
        const leftMostIndex = sides * 2 - rightMostIndex;
        const topMostIndex = 0;
        const bottomMostIndex = Math.floor(sides / 2) * 2;
        const maxX = Math.cos(-import_primitives.TAU + rightMostIndex * step) * w / 2;
        const minX = Math.cos(-import_primitives.TAU + leftMostIndex * step) * w / 2;
        const minY = Math.sin(-import_primitives.TAU + topMostIndex * step) * h / 2;
        const maxY = Math.sin(-import_primitives.TAU + bottomMostIndex * step) * h / 2;
        const diffX = w - Math.abs(maxX - minX);
        const diffY = h - Math.abs(maxY - minY);
        const offsetX = w / 2 + minX - (w / 2 - maxX);
        const offsetY = h / 2 + minY - (h / 2 - maxY);
        const ratio = 1;
        const cx2 = (w - offsetX) / 2;
        const cy2 = (h - offsetY) / 2;
        const ox = (w + diffX) / 2;
        const oy = (h + diffY) / 2;
        const ix = ox * ratio / 2;
        const iy = oy * ratio / 2;
        return Array.from(Array(sides * 2)).map((_, i) => {
          const theta = -import_primitives.TAU + i * step;
          return new import_primitives.Vec2d(
            cx2 + (i % 2 ? ix : ox) * Math.cos(theta),
            cy2 + (i % 2 ? iy : oy) * Math.sin(theta)
          );
        });
      }
      case "rhombus": {
        const offset = Math.min(w * 0.38, h * 0.38);
        return [new import_primitives.Vec2d(offset, 0), new import_primitives.Vec2d(w, 0), new import_primitives.Vec2d(w - offset, h), new import_primitives.Vec2d(0, h)];
      }
      case "rhombus-2": {
        const offset = Math.min(w * 0.38, h * 0.38);
        return [new import_primitives.Vec2d(0, 0), new import_primitives.Vec2d(w - offset, 0), new import_primitives.Vec2d(w, h), new import_primitives.Vec2d(offset, h)];
      }
      case "trapezoid": {
        const offset = Math.min(w * 0.38, h * 0.38);
        return [new import_primitives.Vec2d(offset, 0), new import_primitives.Vec2d(w - offset, 0), new import_primitives.Vec2d(w, h), new import_primitives.Vec2d(0, h)];
      }
      case "arrow-right": {
        const ox = Math.min(w, h) * 0.38;
        const oy = h * 0.16;
        return [
          new import_primitives.Vec2d(0, oy),
          new import_primitives.Vec2d(w - ox, oy),
          new import_primitives.Vec2d(w - ox, 0),
          new import_primitives.Vec2d(w, h / 2),
          new import_primitives.Vec2d(w - ox, h),
          new import_primitives.Vec2d(w - ox, h - oy),
          new import_primitives.Vec2d(0, h - oy)
        ];
      }
      case "arrow-left": {
        const ox = Math.min(w, h) * 0.38;
        const oy = h * 0.16;
        return [
          new import_primitives.Vec2d(ox, 0),
          new import_primitives.Vec2d(ox, oy),
          new import_primitives.Vec2d(w, oy),
          new import_primitives.Vec2d(w, h - oy),
          new import_primitives.Vec2d(ox, h - oy),
          new import_primitives.Vec2d(ox, h),
          new import_primitives.Vec2d(0, h / 2)
        ];
      }
      case "arrow-up": {
        const ox = w * 0.16;
        const oy = Math.min(w, h) * 0.38;
        return [
          new import_primitives.Vec2d(w / 2, 0),
          new import_primitives.Vec2d(w, oy),
          new import_primitives.Vec2d(w - ox, oy),
          new import_primitives.Vec2d(w - ox, h),
          new import_primitives.Vec2d(ox, h),
          new import_primitives.Vec2d(ox, oy),
          new import_primitives.Vec2d(0, oy)
        ];
      }
      case "arrow-down": {
        const ox = w * 0.16;
        const oy = Math.min(w, h) * 0.38;
        return [
          new import_primitives.Vec2d(ox, 0),
          new import_primitives.Vec2d(w - ox, 0),
          new import_primitives.Vec2d(w - ox, h - oy),
          new import_primitives.Vec2d(w, h - oy),
          new import_primitives.Vec2d(w / 2, h),
          new import_primitives.Vec2d(0, h - oy),
          new import_primitives.Vec2d(ox, h - oy)
        ];
      }
      case "check-box":
      case "x-box":
      case "rectangle": {
        return [new import_primitives.Vec2d(0, 0), new import_primitives.Vec2d(w, 0), new import_primitives.Vec2d(w, h), new import_primitives.Vec2d(0, h)];
      }
    }
  }
  onEditEnd = (shape) => {
    const {
      id,
      type,
      props: { text }
    } = shape;
    if (text.trimEnd() !== shape.props.text) {
      this.editor.updateShapes([
        {
          id,
          type,
          props: {
            text: text.trimEnd()
          }
        }
      ]);
    }
  };
  component(shape) {
    const { id, type, props } = shape;
    const forceSolid = (0, import_useForceSolid.useForceSolid)();
    const strokeWidth = import_default_shape_constants.STROKE_SIZES[props.size];
    const { w, color, labelColor, fill, dash, growY, font, align, verticalAlign, size, text } = props;
    const getShape = () => {
      const h = props.h + growY;
      switch (props.geo) {
        case "ellipse": {
          if (dash === "solid" || dash === "draw" && forceSolid) {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SolidStyleEllipse.SolidStyleEllipse, { strokeWidth, w, h, color, fill });
          } else if (dash === "dashed" || dash === "dotted") {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_DashStyleEllipse.DashStyleEllipse,
              {
                id,
                strokeWidth,
                w,
                h,
                dash: dash === "dashed" ? dash : size === "s" && forceSolid ? "dashed" : dash,
                color,
                fill
              }
            );
          } else if (dash === "draw") {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SolidStyleEllipse.SolidStyleEllipse, { strokeWidth, w, h, color, fill });
          }
          break;
        }
        case "oval": {
          if (dash === "solid" || dash === "draw" && forceSolid) {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SolidStyleOval.SolidStyleOval, { strokeWidth, w, h, color, fill });
          } else if (dash === "dashed" || dash === "dotted") {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_DashStyleOval.DashStyleOval,
              {
                id,
                strokeWidth,
                w,
                h,
                dash: dash === "dashed" ? dash : size === "s" && forceSolid ? "dashed" : dash,
                color,
                fill
              }
            );
          } else if (dash === "draw") {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SolidStyleOval.SolidStyleOval, { strokeWidth, w, h, color, fill });
          }
          break;
        }
        default: {
          const outline = this.outline(shape);
          const lines = getLines(shape.props, strokeWidth);
          if (dash === "solid" || dash === "draw" && forceSolid) {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_SolidStylePolygon.SolidStylePolygon,
              {
                fill,
                color,
                strokeWidth,
                outline,
                lines
              }
            );
          } else if (dash === "dashed" || dash === "dotted") {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_DashStylePolygon.DashStylePolygon,
              {
                dash: dash === "dashed" ? dash : size === "s" && forceSolid ? "dashed" : dash,
                fill,
                color,
                strokeWidth,
                outline,
                lines
              }
            );
          } else if (dash === "draw") {
            return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_DrawStylePolygon.DrawStylePolygon,
              {
                id,
                fill,
                color,
                strokeWidth,
                outline,
                lines
              }
            );
          }
        }
      }
    };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_SVGContainer.SVGContainer, { id, children: getShape() }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_TextLabel.TextLabel,
        {
          id,
          type,
          font,
          fill,
          size,
          align,
          verticalAlign,
          text,
          labelColor,
          wrap: true
        }
      ),
      shape.props.url && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_HyperlinkButton.HyperlinkButton, { url: shape.props.url, zoomLevel: this.editor.zoomLevel })
    ] });
  }
  indicator(shape) {
    const { id, props } = shape;
    const { w, h, growY, size } = props;
    const forceSolid = (0, import_useForceSolid.useForceSolid)();
    const strokeWidth = import_default_shape_constants.STROKE_SIZES[size];
    switch (props.geo) {
      case "ellipse": {
        if (props.dash === "draw" && !forceSolid) {
          return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("path", { d: (0, import_DrawStyleEllipse.getEllipseIndicatorPath)(id, w, h + growY, strokeWidth) });
        }
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("ellipse", { cx: w / 2, cy: (h + growY) / 2, rx: w / 2, ry: (h + growY) / 2 });
      }
      case "oval": {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("path", { d: (0, import_SolidStyleOval.getOvalIndicatorPath)(w, h + growY) });
      }
      default: {
        const outline = this.outline(shape);
        let path;
        if (props.dash === "draw" && !forceSolid) {
          const polygonPoints = (0, import_primitives.getRoundedPolygonPoints)(id, outline, 0, strokeWidth * 2, 1);
          path = (0, import_primitives.getRoundedInkyPolygonPath)(polygonPoints);
        } else {
          path = "M" + outline[0] + "L" + outline.slice(1) + "Z";
        }
        const lines = getLines(shape.props, strokeWidth);
        if (lines) {
          for (const [A, B] of lines) {
            path += `M${A.x},${A.y}L${B.x},${B.y}`;
          }
        }
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("path", { d: path });
      }
    }
  }
  toSvg(shape, font, colors) {
    const { id, props } = shape;
    const strokeWidth = import_default_shape_constants.STROKE_SIZES[props.size];
    let svgElm;
    switch (props.geo) {
      case "ellipse": {
        switch (props.dash) {
          case "draw":
            svgElm = (0, import_DrawStyleEllipse.DrawStyleEllipseSvg)({
              id,
              w: props.w,
              h: props.h,
              color: props.color,
              fill: props.fill,
              strokeWidth,
              colors
            });
            break;
          case "solid":
            svgElm = (0, import_SolidStyleEllipse.SolidStyleEllipseSvg)({
              strokeWidth,
              w: props.w,
              h: props.h,
              color: props.color,
              fill: props.fill,
              colors
            });
            break;
          default:
            svgElm = (0, import_DashStyleEllipse.DashStyleEllipseSvg)({
              id,
              strokeWidth,
              w: props.w,
              h: props.h,
              dash: props.dash,
              color: props.color,
              fill: props.fill,
              colors
            });
            break;
        }
        break;
      }
      case "oval": {
        switch (props.dash) {
          case "draw":
            svgElm = (0, import_DashStyleOval.DashStyleOvalSvg)({
              id,
              strokeWidth,
              w: props.w,
              h: props.h,
              dash: props.dash,
              color: props.color,
              fill: props.fill,
              colors
            });
            break;
          case "solid":
            svgElm = (0, import_SolidStyleOval.SolidStyleOvalSvg)({
              strokeWidth,
              w: props.w,
              h: props.h,
              color: props.color,
              fill: props.fill,
              colors
            });
            break;
          default:
            svgElm = (0, import_DashStyleOval.DashStyleOvalSvg)({
              id,
              strokeWidth,
              w: props.w,
              h: props.h,
              dash: props.dash,
              color: props.color,
              fill: props.fill,
              colors
            });
        }
        break;
      }
      default: {
        const outline = this.outline(shape);
        const lines = getLines(shape.props, strokeWidth);
        switch (props.dash) {
          case "draw":
            svgElm = (0, import_DrawStylePolygon.DrawStylePolygonSvg)({
              id,
              fill: props.fill,
              color: props.color,
              strokeWidth,
              outline,
              lines,
              colors
            });
            break;
          case "solid":
            svgElm = (0, import_SolidStylePolygon.SolidStylePolygonSvg)({
              fill: props.fill,
              color: props.color,
              strokeWidth,
              outline,
              lines,
              colors
            });
            break;
          default:
            svgElm = (0, import_DashStylePolygon.DashStylePolygonSvg)({
              dash: props.dash,
              fill: props.fill,
              color: props.color,
              strokeWidth,
              outline,
              lines,
              colors
            });
            break;
        }
        break;
      }
    }
    if (props.text) {
      const bounds = this.bounds(shape);
      const rootTextElm = (0, import_getTextLabelSvgElement.getTextLabelSvgElement)({
        editor: this.editor,
        shape,
        font,
        bounds
      });
      const textElm = rootTextElm.cloneNode(true);
      textElm.setAttribute("fill", colors.fill[shape.props.labelColor]);
      textElm.setAttribute("stroke", "none");
      const textBgEl = rootTextElm.cloneNode(true);
      textBgEl.setAttribute("stroke-width", "2");
      textBgEl.setAttribute("fill", colors.background);
      textBgEl.setAttribute("stroke", colors.background);
      const groupEl = document.createElementNS("http://www.w3.org/2000/svg", "g");
      groupEl.append(textBgEl);
      groupEl.append(textElm);
      if (svgElm.nodeName === "g") {
        svgElm.appendChild(groupEl);
        return svgElm;
      } else {
        const g = document.createElementNS("http://www.w3.org/2000/svg", "g");
        g.appendChild(svgElm);
        g.appendChild(groupEl);
        return g;
      }
    }
    return svgElm;
  }
  onResize = (shape, { initialBounds, handle, newPoint, scaleX, scaleY }) => {
    let w = initialBounds.width * scaleX;
    let h = initialBounds.height * scaleY;
    let overShrinkX = 0;
    let overShrinkY = 0;
    if (shape.props.text.trim()) {
      let newW = Math.max(Math.abs(w), MIN_SIZE_WITH_LABEL);
      let newH = Math.max(Math.abs(h), MIN_SIZE_WITH_LABEL);
      if (newW < MIN_SIZE_WITH_LABEL && newH === MIN_SIZE_WITH_LABEL) {
        newW = MIN_SIZE_WITH_LABEL;
      }
      if (newW === MIN_SIZE_WITH_LABEL && newH < MIN_SIZE_WITH_LABEL) {
        newH = MIN_SIZE_WITH_LABEL;
      }
      const labelSize = getLabelSize(this.editor, {
        ...shape,
        props: {
          ...shape.props,
          w: newW,
          h: newH
        }
      });
      const nextW = Math.max(Math.abs(w), labelSize.w) * Math.sign(w);
      const nextH = Math.max(Math.abs(h), labelSize.h) * Math.sign(h);
      overShrinkX = Math.abs(nextW) - Math.abs(w);
      overShrinkY = Math.abs(nextH) - Math.abs(h);
      w = nextW;
      h = nextH;
    }
    const offset = new import_primitives.Vec2d(0, 0);
    if (scaleX < 0) {
      offset.x += w;
    }
    if (handle === "left" || handle === "top_left" || handle === "bottom_left") {
      offset.x += scaleX < 0 ? overShrinkX : -overShrinkX;
    }
    if (scaleY < 0) {
      offset.y += h;
    }
    if (handle === "top" || handle === "top_left" || handle === "top_right") {
      offset.y += scaleY < 0 ? overShrinkY : -overShrinkY;
    }
    const { x, y } = offset.rot(shape.rotation).add(newPoint);
    return {
      x,
      y,
      props: {
        w: Math.max(Math.abs(w), 1),
        h: Math.max(Math.abs(h), 1),
        growY: 0
      }
    };
  };
  onBeforeCreate = (shape) => {
    if (!shape.props.text) {
      if (shape.props.growY) {
        return {
          ...shape,
          props: {
            ...shape.props,
            growY: 0
          }
        };
      } else {
        return;
      }
    }
    const prevHeight = shape.props.h;
    const nextHeight = getLabelSize(this.editor, shape).h;
    let growY = null;
    if (nextHeight > prevHeight) {
      growY = nextHeight - prevHeight;
    } else {
      if (shape.props.growY) {
        growY = 0;
      }
    }
    if (growY !== null) {
      return {
        ...shape,
        props: {
          ...shape.props,
          growY
        }
      };
    }
  };
  onBeforeUpdate = (prev, next) => {
    const prevText = prev.props.text.trimEnd();
    const nextText = next.props.text.trimEnd();
    if (prevText === nextText && prev.props.font === next.props.font && prev.props.size === next.props.size) {
      return;
    }
    if (prevText && !nextText) {
      return {
        ...next,
        props: {
          ...next.props,
          growY: 0
        }
      };
    }
    const prevWidth = prev.props.w;
    const prevHeight = prev.props.h;
    const nextSize = getLabelSize(this.editor, next);
    const nextWidth = nextSize.w;
    const nextHeight = nextSize.h;
    if (!prevText && nextText && nextText.length === 1) {
      let w = Math.max(prevWidth, nextWidth);
      let h = Math.max(prevHeight, nextHeight);
      if (prev.props.w < MIN_SIZE_WITH_LABEL && prev.props.h < MIN_SIZE_WITH_LABEL) {
        w = Math.max(w, MIN_SIZE_WITH_LABEL);
        h = Math.max(h, MIN_SIZE_WITH_LABEL);
        w = Math.max(w, h);
        h = Math.max(w, h);
      }
      return {
        ...next,
        props: {
          ...next.props,
          w,
          h,
          growY: 0
        }
      };
    }
    let growY = null;
    if (nextHeight > prevHeight) {
      growY = nextHeight - prevHeight;
    } else {
      if (prev.props.growY) {
        growY = 0;
      }
    }
    if (growY !== null) {
      return {
        ...next,
        props: {
          ...next.props,
          growY,
          w: Math.max(next.props.w, nextWidth)
        }
      };
    }
    if (nextWidth > prev.props.w) {
      return {
        ...next,
        props: {
          ...next.props,
          w: nextWidth
        }
      };
    }
  };
  onDoubleClick = (shape) => {
    if (this.editor.inputs.altKey) {
      switch (shape.props.geo) {
        case "rectangle": {
          return {
            ...shape,
            props: {
              geo: "check-box"
            }
          };
        }
        case "check-box": {
          return {
            ...shape,
            props: {
              geo: "rectangle"
            }
          };
        }
      }
    }
    return;
  };
}
function getLabelSize(editor, shape) {
  const text = shape.props.text.trimEnd();
  if (!text) {
    return { w: 0, h: 0 };
  }
  const minSize = editor.textMeasure.measureText("w", {
    ...import_default_shape_constants.TEXT_PROPS,
    fontFamily: import_default_shape_constants.FONT_FAMILIES[shape.props.font],
    fontSize: import_default_shape_constants.LABEL_FONT_SIZES[shape.props.size],
    width: "fit-content",
    maxWidth: "100px"
  });
  const sizes = {
    s: 2,
    m: 3.5,
    l: 5,
    xl: 10
  };
  const size = editor.textMeasure.measureText(text, {
    ...import_default_shape_constants.TEXT_PROPS,
    fontFamily: import_default_shape_constants.FONT_FAMILIES[shape.props.font],
    fontSize: import_default_shape_constants.LABEL_FONT_SIZES[shape.props.size],
    width: "fit-content",
    minWidth: minSize.w + "px",
    maxWidth: Math.max(
      // Guard because a DOM nodes can't be less 0
      0,
      // A 'w' width that we're setting as the min-width
      Math.ceil(minSize.w + sizes[shape.props.size]),
      // The actual text size
      Math.ceil(shape.props.w - LABEL_PADDING * 2)
    ) + "px"
  });
  return {
    w: size.w + LABEL_PADDING * 2,
    h: size.h + LABEL_PADDING * 2
  };
}
function getLines(props, sw) {
  switch (props.geo) {
    case "x-box": {
      return getXBoxLines(props.w, props.h, sw, props.dash);
    }
    case "check-box": {
      return getCheckBoxLines(props.w, props.h);
    }
    default: {
      return void 0;
    }
  }
}
function getXBoxLines(w, h, sw, dash) {
  const inset = dash === "draw" ? 0.62 : 0;
  if (dash === "dashed") {
    return [
      [new import_primitives.Vec2d(0, 0), new import_primitives.Vec2d(w / 2, h / 2)],
      [new import_primitives.Vec2d(w, h), new import_primitives.Vec2d(w / 2, h / 2)],
      [new import_primitives.Vec2d(0, h), new import_primitives.Vec2d(w / 2, h / 2)],
      [new import_primitives.Vec2d(w, 0), new import_primitives.Vec2d(w / 2, h / 2)]
    ];
  }
  return [
    [new import_primitives.Vec2d(sw * inset, sw * inset), new import_primitives.Vec2d(w - sw * inset, h - sw * inset)],
    [new import_primitives.Vec2d(sw * inset, h - sw * inset), new import_primitives.Vec2d(w - sw * inset, sw * inset)]
  ];
}
function getCheckBoxLines(w, h) {
  const size = Math.min(w, h) * 0.82;
  const ox = (w - size) / 2;
  const oy = (h - size) / 2;
  return [
    [new import_primitives.Vec2d(ox + size * 0.25, oy + size * 0.52), new import_primitives.Vec2d(ox + size * 0.45, oy + size * 0.82)],
    [new import_primitives.Vec2d(ox + size * 0.45, oy + size * 0.82), new import_primitives.Vec2d(ox + size * 0.82, oy + size * 0.22)]
  ];
}
