// @tldraw/ui/dist-cjs/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  AssetUrlsProvider: () => import_useAssetUrls.AssetUrlsProvider,
  BreakPointProvider: () => import_useBreakpoint.BreakPointProvider,
  Button: () => import_Button.Button,
  ContextMenu: () => import_ContextMenu.ContextMenu,
  Dialog: () => Dialog,
  DropdownMenu: () => DropdownMenu,
  Icon: () => import_Icon.Icon,
  Input: () => import_Input.Input,
  TldrawUi: () => import_TldrawUi.TldrawUi,
  TldrawUiContextProvider: () => import_TldrawUiContextProvider.TldrawUiContextProvider,
  compactMenuItems: () => import_menuHelpers.compactMenuItems,
  findMenuItem: () => import_menuHelpers.findMenuItem,
  menuCustom: () => import_menuHelpers.menuCustom,
  menuGroup: () => import_menuHelpers.menuGroup,
  menuItem: () => import_menuHelpers.menuItem,
  menuSubmenu: () => import_menuHelpers.menuSubmenu,
  setDefaultUiAssetUrls: () => import_assetUrls.setDefaultUiAssetUrls,
  toolbarItem: () => import_useToolbarSchema.toolbarItem,
  useActions: () => import_useActions.useActions,
  useActionsMenuSchema: () => import_useActionsMenuSchema.useActionsMenuSchema,
  useAssetUrls: () => import_useAssetUrls.useAssetUrls,
  useBreakpoint: () => import_useBreakpoint.useBreakpoint,
  useCanRedo: () => import_useCanRedo.useCanRedo,
  useCanUndo: () => import_useCanUndo.useCanUndo,
  useContextMenuSchema: () => import_useContextMenuSchema.useContextMenuSchema,
  useCopyAs: () => import_useCopyAs.useCopyAs,
  useDefaultHelpers: () => import_overrides.useDefaultHelpers,
  useDialogs: () => import_useDialogsProvider.useDialogs,
  useEvents: () => import_useEventsProvider.useEvents,
  useExportAs: () => import_useExportAs.useExportAs,
  useHelpMenuSchema: () => import_useHelpMenuSchema.useHelpMenuSchema,
  useKeyboardShortcuts: () => import_useKeyboardShortcuts.useKeyboardShortcuts,
  useKeyboardShortcutsSchema: () => import_useKeyboardShortcutsSchema.useKeyboardShortcutsSchema,
  useLocalStorageState: () => import_useLocalStorageState.useLocalStorageState,
  useMenuClipboardEvents: () => import_useClipboardEvents.useMenuClipboardEvents,
  useMenuIsOpen: () => import_useMenuIsOpen.useMenuIsOpen,
  useMenuSchema: () => import_useMenuSchema.useMenuSchema,
  useNativeClipboardEvents: () => import_useClipboardEvents.useNativeClipboardEvents,
  useReadonly: () => import_useReadonly.useReadonly,
  useToasts: () => import_useToastsProvider.useToasts,
  useToolbarSchema: () => import_useToolbarSchema.useToolbarSchema,
  useTools: () => import_useTools.useTools,
  useTranslation: () => import_useTranslation.useTranslation
});
module.exports = __toCommonJS(src_exports);
var Dialog = __toESM(require("./lib/components/primitives/Dialog"));
var DropdownMenu = __toESM(require("./lib/components/primitives/DropdownMenu"));
var import_TldrawUi = require("./lib/TldrawUi");
var import_TldrawUiContextProvider = require("./lib/TldrawUiContextProvider");
var import_assetUrls = require("./lib/assetUrls");
var import_ContextMenu = require("./lib/components/ContextMenu");
var import_Button = require("./lib/components/primitives/Button");
var import_Icon = require("./lib/components/primitives/Icon");
var import_Input = require("./lib/components/primitives/Input");
var import_menuHelpers = require("./lib/hooks/menuHelpers");
var import_useActions = require("./lib/hooks/useActions");
var import_useActionsMenuSchema = require("./lib/hooks/useActionsMenuSchema");
var import_useAssetUrls = require("./lib/hooks/useAssetUrls");
var import_useBreakpoint = require("./lib/hooks/useBreakpoint");
var import_useCanRedo = require("./lib/hooks/useCanRedo");
var import_useCanUndo = require("./lib/hooks/useCanUndo");
var import_useClipboardEvents = require("./lib/hooks/useClipboardEvents");
var import_useContextMenuSchema = require("./lib/hooks/useContextMenuSchema");
var import_useCopyAs = require("./lib/hooks/useCopyAs");
var import_useDialogsProvider = require("./lib/hooks/useDialogsProvider");
var import_useEventsProvider = require("./lib/hooks/useEventsProvider");
var import_useExportAs = require("./lib/hooks/useExportAs");
var import_useHelpMenuSchema = require("./lib/hooks/useHelpMenuSchema");
var import_useKeyboardShortcuts = require("./lib/hooks/useKeyboardShortcuts");
var import_useKeyboardShortcutsSchema = require("./lib/hooks/useKeyboardShortcutsSchema");
var import_useLocalStorageState = require("./lib/hooks/useLocalStorageState");
var import_useMenuIsOpen = require("./lib/hooks/useMenuIsOpen");
var import_useMenuSchema = require("./lib/hooks/useMenuSchema");
var import_useReadonly = require("./lib/hooks/useReadonly");
var import_useToastsProvider = require("./lib/hooks/useToastsProvider");
var import_useToolbarSchema = require("./lib/hooks/useToolbarSchema");
var import_useTools = require("./lib/hooks/useTools");
var import_useTranslation = require("./lib/hooks/useTranslation/useTranslation");
var import_overrides = require("./lib/overrides");
