// @tldraw/editor/dist-cjs/lib/editor/managers/DragAndDropManager.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DragAndDropManager_exports = {};
__export(DragAndDropManager_exports, {
  DragAndDropManager: () => DragAndDropManager
});
module.exports = __toCommonJS(DragAndDropManager_exports);
var import_utils = require("@tldraw/utils");
const LAG_DURATION = 100;
class DragAndDropManager {
  constructor(editor) {
    this.editor = editor;
    editor.disposables.add(this.dispose);
  }
  prevDroppingShapeId = null;
  currDroppingShapeId = null;
  droppingNodeTimer = null;
  updateDroppingNode(movingShapes, cb) {
    if (this.droppingNodeTimer === null) {
      const { currentPagePoint } = this.editor.inputs;
      this.currDroppingShapeId = this.editor.getDroppingShape(currentPagePoint, movingShapes)?.id ?? null;
      this.setDragTimer(movingShapes, LAG_DURATION * 10, cb);
    } else if (this.editor.inputs.pointerVelocity.len() > 0.5) {
      clearInterval(this.droppingNodeTimer);
      this.setDragTimer(movingShapes, LAG_DURATION, cb);
    }
  }
  setDragTimer(movingShapes, duration, cb) {
    this.droppingNodeTimer = setTimeout(() => {
      this.editor.batch(() => {
        this.handleDrag(movingShapes, cb);
      });
      this.droppingNodeTimer = null;
    }, duration);
  }
  handleDrag(movingShapes, cb) {
    const { currentPagePoint } = this.editor.inputs;
    movingShapes = (0, import_utils.compact)(movingShapes.map((shape) => this.editor.getShapeById(shape.id)));
    const currDroppingShapeId = this.editor.getDroppingShape(currentPagePoint, movingShapes)?.id ?? null;
    if (currDroppingShapeId !== this.currDroppingShapeId) {
      this.prevDroppingShapeId = this.currDroppingShapeId;
      this.currDroppingShapeId = currDroppingShapeId;
    }
    const { prevDroppingShapeId } = this;
    if (currDroppingShapeId === prevDroppingShapeId) {
      return;
    }
    const prevDroppingShape = prevDroppingShapeId && this.editor.getShapeById(prevDroppingShapeId);
    const nextDroppingShape = currDroppingShapeId && this.editor.getShapeById(currDroppingShapeId);
    this.prevDroppingShapeId = this.currDroppingShapeId;
    if (prevDroppingShape) {
      this.editor.getShapeUtil(prevDroppingShape).onDragShapesOut?.(prevDroppingShape, movingShapes);
    }
    if (nextDroppingShape) {
      const res = this.editor.getShapeUtil(nextDroppingShape).onDragShapesOver?.(nextDroppingShape, movingShapes);
      if (res && res.shouldHint) {
        this.editor.setHintingIds([nextDroppingShape.id]);
      }
    } else {
      this.editor.setHintingIds([]);
    }
    cb?.();
  }
  dropShapes(shapes) {
    const { currDroppingShapeId } = this;
    this.handleDrag(shapes);
    if (currDroppingShapeId) {
      const shape = this.editor.getShapeById(currDroppingShapeId);
      if (!shape)
        return;
      this.editor.getShapeUtil(shape).onDropShapesOver?.(shape, shapes);
    }
  }
  clear() {
    this.prevDroppingShapeId = null;
    this.currDroppingShapeId = null;
    if (this.droppingNodeTimer !== null) {
      clearInterval(this.droppingNodeTimer);
    }
    this.droppingNodeTimer = null;
    this.editor.setHintingIds([]);
  }
  dispose = () => {
    this.clear();
  };
}
