// y-partykit/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  onConnect: () => onConnect
});
module.exports = __toCommonJS(src_exports);
var Y2 = __toESM(require("yjs"));
var syncProtocol = __toESM(require("y-protocols/sync"));
var awarenessProtocol = __toESM(require("y-protocols/awareness"));
var encoding2 = __toESM(require("lib0/encoding"));
var decoding2 = __toESM(require("lib0/decoding"));
var import_lodash = __toESM(require("lodash.debounce"));

// src/storage.ts
var Y = __toESM(require("yjs"));
var encoding = __toESM(require("lib0/encoding"));
var decoding = __toESM(require("lib0/decoding"));
var PREFERRED_TRIM_SIZE = 300;
var BINARY_BITS_32 = 4294967295;
var keyEncoding = {
  encode(arr) {
    const resultArr = [];
    for (const item of arr) {
      resultArr.push(
        // TODO: This is a bit hacky, but it works
        typeof item === "string" ? `"${item}"` : `${item}`.padStart(9, "0")
      );
    }
    return resultArr.join("#");
  },
  decode(str) {
    return str.split("#").map(
      (el) => el.startsWith('"') ? JSON.parse(el) : parseInt(el, 10)
    );
  }
};
async function levelGet(db, key) {
  const prefix = keyEncoding.encode(key);
  const res = await db.list({
    start: prefix,
    end: `${prefix}#zzzzz`
  });
  if (res.size === 0) {
    return null;
  }
  const finalArrayLength = Array.from(res.values()).reduce(
    (acc, val) => acc + val.length,
    0
  );
  const finalArray = new Uint8Array(finalArrayLength);
  let offset = 0;
  for (const val of res.values()) {
    finalArray.set(val, offset);
    offset += val.length;
  }
  return finalArray;
}
async function levelPut(db, key, val) {
  const chunks = [];
  for (let i = 0; i < val.length; i += 128 * 1024) {
    chunks.push(val.slice(i, i + 128 * 1024));
  }
  const keyPrefix = keyEncoding.encode(key);
  for (let i = 0; i < chunks.length; i++) {
    await db.put(`${keyPrefix}#${i}`, chunks[i]);
  }
}
function groupBy(arr, fn) {
  const map = /* @__PURE__ */ new Map();
  for (const el of arr) {
    const key = fn(el);
    if (!map.has(key)) {
      map.set(key, []);
    }
    map.get(key).push(el);
  }
  return map;
}
async function getLevelBulkData(db, opts) {
  const res = await db.list({
    start: keyEncoding.encode(opts.gte),
    end: keyEncoding.encode(opts.lt),
    reverse: opts.reverse,
    limit: opts.limit
  });
  const grouped = groupBy(
    Array.from(res.entries()),
    ([key]) => key.split("#").slice(0, -1).join("#")
  );
  const result = [];
  for (const [key, values] of grouped.entries()) {
    const finalArrayLength = values.reduce(
      (acc, val) => acc + val[1].length,
      0
    );
    const finalArray = new Uint8Array(finalArrayLength);
    let offset = 0;
    for (const [, val] of values) {
      finalArray.set(val, offset);
      offset += val.length;
    }
    result.push({
      key: keyEncoding.decode(key),
      value: finalArray
    });
  }
  return result;
}
async function getLevelUpdates(db, docName, opts = {
  values: true,
  keys: false
}) {
  return getLevelBulkData(db, {
    gte: createDocumentUpdateKey(docName, 0),
    lt: createDocumentUpdateKey(docName, BINARY_BITS_32),
    ...opts
  });
}
async function getCurrentUpdateClock(db, docName) {
  return getLevelUpdates(db, docName, {
    keys: true,
    values: false,
    reverse: true,
    limit: 1
  }).then((datums) => {
    if (datums.length === 0) {
      return -1;
    } else {
      const ret = datums[0].key[3];
      if (typeof ret !== "number") {
        throw new Error("Expected number, got " + typeof ret);
      }
      return ret;
    }
  });
}
async function clearRange(db, gte, lt) {
  const datums = await getLevelBulkData(db, {
    values: false,
    keys: true,
    gte,
    lt
  });
  const arr = [];
  for (const [index, datum] of datums.entries()) {
    arr.push(keyEncoding.encode(datum.key));
    if (arr.length === 128 || index === datums.length - 1) {
      await db.delete(arr);
      arr.length = 0;
    }
  }
}
async function clearUpdatesRange(db, docName, from, to) {
  return clearRange(
    db,
    createDocumentUpdateKey(docName, from),
    createDocumentUpdateKey(docName, to)
  );
}
function createDocumentUpdateKey(docName, clock) {
  return ["v1", docName, "update", clock];
}
function createDocumentStateVectorKey(docName) {
  return ["v1_sv", docName];
}
function mergeUpdates(updates) {
  const ydoc = new Y.Doc();
  ydoc.transact(() => {
    for (let i = 0; i < updates.length; i++) {
      Y.applyUpdate(ydoc, updates[i]);
    }
  });
  return { update: Y.encodeStateAsUpdate(ydoc), sv: Y.encodeStateVector(ydoc) };
}
async function writeStateVector(db, docName, sv, clock) {
  const encoder = encoding.createEncoder();
  encoding.writeVarUint(encoder, clock);
  encoding.writeVarUint8Array(encoder, sv);
  await levelPut(
    db,
    createDocumentStateVectorKey(docName),
    encoding.toUint8Array(encoder)
  );
}
function decodeLeveldbStateVector(buf) {
  const decoder = decoding.createDecoder(buf);
  const clock = decoding.readVarUint(decoder);
  const sv = decoding.readVarUint8Array(decoder);
  return { sv, clock };
}
async function readStateVector(db, docName) {
  const buf = await levelGet(db, createDocumentStateVectorKey(docName));
  if (buf === null) {
    return { sv: null, clock: -1 };
  }
  return decodeLeveldbStateVector(buf);
}
async function flushDocument(db, docName, stateAsUpdate, stateVector) {
  const clock = await storeUpdate(db, docName, stateAsUpdate);
  await writeStateVector(db, docName, stateVector, clock);
  await clearUpdatesRange(db, docName, 0, clock);
  return clock;
}
async function storeUpdate(db, docName, update) {
  const clock = await getCurrentUpdateClock(db, docName);
  if (clock === -1) {
    const ydoc = new Y.Doc();
    Y.applyUpdate(ydoc, update);
    const sv = Y.encodeStateVector(ydoc);
    await writeStateVector(db, docName, sv, 0);
  }
  await levelPut(db, createDocumentUpdateKey(docName, clock + 1), update);
  return clock + 1;
}
var YPartyKitStorage = class {
  db;
  tr;
  _transact(fn) {
    throw Error("implement _transact");
    return fn(this.db);
  }
  constructor(storage) {
    const db = this.db = storage;
    this.tr = Promise.resolve();
    this._transact = (f) => {
      const currTr = this.tr;
      this.tr = (async () => {
        await currTr;
        let res = (
          /** @type {any} */
          null
        );
        try {
          res = await f(db);
        } catch (err) {
          console.warn("Error during y-partykit-storage transaction", err);
        }
        return res;
      })();
      return this.tr;
    };
  }
  async flushDocument(docName) {
    return this._transact(async (db) => {
      const updates = await getLevelUpdates(db, docName);
      const { update, sv } = mergeUpdates(updates.map((u) => u.value));
      await flushDocument(db, docName, update, sv);
    });
  }
  async getYDoc(docName) {
    return this._transact(async (db) => {
      const updates = await getLevelUpdates(db, docName);
      const ydoc = new Y.Doc();
      ydoc.transact(() => {
        for (let i = 0; i < updates.length; i++) {
          Y.applyUpdate(ydoc, updates[i].value);
        }
      });
      if (updates.length > PREFERRED_TRIM_SIZE) {
        await flushDocument(
          db,
          docName,
          Y.encodeStateAsUpdate(ydoc),
          Y.encodeStateVector(ydoc)
        );
      }
      return ydoc;
    });
  }
  async getStateVector(docName) {
    return this._transact(async (db) => {
      const { clock, sv } = await readStateVector(db, docName);
      let curClock = -1;
      if (sv !== null) {
        curClock = await getCurrentUpdateClock(db, docName);
      }
      if (sv !== null && clock === curClock) {
        return sv;
      } else {
        const updates = await getLevelUpdates(db, docName);
        const { update, sv: sv2 } = mergeUpdates(updates.map((u) => u.value));
        await flushDocument(db, docName, update, sv2);
        return sv2;
      }
    });
  }
  async storeUpdate(docName, update) {
    return this._transact((db) => storeUpdate(db, docName, update));
  }
};

// src/index.ts
function assert(condition, message) {
  if (!condition) {
    throw new Error(message);
  }
}
var wsReadyStateConnecting = 0;
var wsReadyStateOpen = 1;
var docs = /* @__PURE__ */ new Map();
var messageSync = 0;
var messageAwareness = 1;
function updateHandler(update, origin, doc) {
  const encoder = encoding2.createEncoder();
  encoding2.writeVarUint(encoder, messageSync);
  syncProtocol.writeUpdate(encoder, update);
  const message = encoding2.toUint8Array(encoder);
  doc.conns.forEach((_, conn) => send(doc, conn, message));
}
var WSSharedDoc = class extends Y2.Doc {
  name;
  conns;
  awareness;
  storage;
  persist;
  constructor(room, options) {
    super({ gc: options.gc || false });
    this.name = room.id;
    this.persist = options.persist || false;
    if (options.persist) {
      this.storage = new YPartyKitStorage(room.storage);
    }
    this.conns = /* @__PURE__ */ new Map();
    this.awareness = new awarenessProtocol.Awareness(this);
    this.awareness.setLocalState(null);
    const awarenessChangeHandler = ({
      added,
      updated,
      removed
    }, conn) => {
      const changedClients = added.concat(updated, removed);
      if (conn !== null) {
        const connControlledIDs = (
          /** @type {Set<number>} */
          this.conns.get(conn)
        );
        if (connControlledIDs !== void 0) {
          added.forEach((clientID) => {
            connControlledIDs.add(clientID);
          });
          removed.forEach((clientID) => {
            connControlledIDs.delete(clientID);
          });
        }
      }
      const encoder = encoding2.createEncoder();
      encoding2.writeVarUint(encoder, messageAwareness);
      encoding2.writeVarUint8Array(
        encoder,
        awarenessProtocol.encodeAwarenessUpdate(this.awareness, changedClients)
      );
      const buff = encoding2.toUint8Array(encoder);
      this.conns.forEach((_, c) => {
        send(this, c, buff);
      });
    };
    this.awareness.on("update", awarenessChangeHandler);
    this.on("update", updateHandler);
  }
  async bindState() {
    assert(this.storage, "Storage not set");
    const persistedYdoc = await this.storage.getYDoc(this.name);
    const newUpdates = Y2.encodeStateAsUpdate(this);
    await this.storage.storeUpdate(this.name, newUpdates);
    Y2.applyUpdate(this, Y2.encodeStateAsUpdate(persistedYdoc));
    this.on("update", (update) => {
      assert(this.storage, "Storage not set");
      this.storage.storeUpdate(this.name, update).catch((e) => {
        console.error("Error storing update", e);
      });
    });
  }
  async writeState() {
    assert(this.storage, "Storage not set");
    const newUpdates = Y2.encodeStateAsUpdate(this);
    await this.storage.storeUpdate(this.name, newUpdates);
  }
};
var CALLBACK_DEFAULTS = {
  debounceWait: 2e3,
  debounceMaxWait: 1e4,
  timeout: 5e3,
  objects: {}
};
function getContent(objName, objType, doc) {
  switch (objType) {
    case "Array":
      return doc.getArray(objName);
    case "Map":
      return doc.getMap(objName);
    case "Text":
      return doc.getText(objName);
    case "XmlFragment":
      return doc.getXmlFragment(objName);
    case "XmlElement":
      return doc.getXmlElement(objName);
    default:
      return {};
  }
}
async function getYDoc(room, options) {
  let doc = docs.get(room.id);
  if (doc) {
    return doc;
  }
  doc = new WSSharedDoc(room, options);
  const { callback, load } = options;
  if (load) {
    const src = await load();
    const state = Y2.encodeStateAsUpdate(src);
    Y2.applyUpdate(doc, state);
  }
  doc.gc = options.gc || false;
  if (callback !== void 0) {
    doc.on(
      "update",
      (0, import_lodash.default)(
        (update, origin, doc2) => {
          const dataToSend = {
            room: doc2.name,
            data: {}
          };
          const callbackObjects = callback.objects || CALLBACK_DEFAULTS.objects;
          const sharedObjectList = Object.keys(callbackObjects);
          sharedObjectList.forEach((sharedObjectName) => {
            const sharedObjectType = callbackObjects[sharedObjectName];
            dataToSend.data[sharedObjectName] = {
              type: sharedObjectType,
              content: getContent(
                sharedObjectName,
                sharedObjectType,
                doc2
              ).toJSON()
            };
          });
          if (callback.url) {
            fetch(callback.url, {
              method: "POST",
              headers: {
                "Content-Type": "application/json"
              },
              body: JSON.stringify(dataToSend),
              signal: AbortSignal.timeout(
                callback.timeout || CALLBACK_DEFAULTS.timeout
              )
            }).catch((err) => {
              console.error("failed to persist", err);
            });
          }
          if (callback.handler) {
            callback.handler(doc2);
          }
        },
        callback.debounceWait || CALLBACK_DEFAULTS.debounceWait,
        {
          maxWait: callback.debounceMaxWait || CALLBACK_DEFAULTS.debounceMaxWait
        }
      )
    );
  }
  if (doc.persist) {
    await doc.bindState();
  }
  docs.set(room.id, doc);
  return doc;
}
function messageListener(conn, doc, message) {
  try {
    const encoder = encoding2.createEncoder();
    const decoder = decoding2.createDecoder(message);
    const messageType = decoding2.readVarUint(decoder);
    switch (messageType) {
      case messageSync:
        encoding2.writeVarUint(encoder, messageSync);
        syncProtocol.readSyncMessage(decoder, encoder, doc, conn);
        if (encoding2.length(encoder) > 1) {
          send(doc, conn, encoding2.toUint8Array(encoder));
        }
        break;
      case messageAwareness: {
        awarenessProtocol.applyAwarenessUpdate(
          doc.awareness,
          decoding2.readVarUint8Array(decoder),
          conn
        );
        break;
      }
    }
  } catch (err) {
    console.error(err);
    doc.emit("error", [err]);
  }
}
function closeConn(doc, conn) {
  if (doc.conns.has(conn)) {
    const controlledIds = doc.conns.get(conn);
    doc.conns.delete(conn);
    awarenessProtocol.removeAwarenessStates(
      doc.awareness,
      Array.from(controlledIds),
      null
    );
    if (doc.conns.size === 0 && doc.persist) {
      doc.writeState().then(
        () => {
          doc.destroy();
        },
        (err) => {
          doc.emit("error", [err]);
        }
      );
      docs.delete(doc.name);
    }
  }
  try {
    conn.close();
  } catch (e) {
    console.warn("failed to close connection", e);
  }
}
function send(doc, conn, m) {
  if (conn.readyState !== void 0 && conn.readyState !== wsReadyStateConnecting && conn.readyState !== wsReadyStateOpen) {
    closeConn(doc, conn);
  }
  try {
    conn.send(
      m
      // /** @param {any} err */ (err) => {
      //   err != null && closeConn(doc, conn);
      // }
    );
  } catch (e) {
    closeConn(doc, conn);
  }
}
var pingTimeout = 3e4;
async function onConnect(conn, room, options = {}) {
  const doc = await getYDoc(room, options);
  doc.conns.set(conn, /* @__PURE__ */ new Set());
  conn.addEventListener("message", (message) => {
    if (typeof message.data !== "string") {
      return messageListener(conn, doc, new Uint8Array(message.data));
    } else {
    }
  });
  let pongReceived = true;
  const pingInterval = setInterval(() => {
    if (!pongReceived) {
      if (doc.conns.has(conn)) {
        closeConn(doc, conn);
      }
      clearInterval(pingInterval);
    } else if (doc.conns.has(conn)) {
      pongReceived = false;
      try {
        conn.send("ping");
      } catch (e) {
        closeConn(doc, conn);
        clearInterval(pingInterval);
      }
    }
  }, pingTimeout);
  conn.addEventListener("close", () => {
    closeConn(doc, conn);
    clearInterval(pingInterval);
  });
  conn.addEventListener("message", (message) => {
    if (message.data === "pong") {
      pongReceived = true;
    }
  });
  {
    const encoder = encoding2.createEncoder();
    encoding2.writeVarUint(encoder, messageSync);
    syncProtocol.writeSyncStep1(encoder, doc);
    send(doc, conn, encoding2.toUint8Array(encoder));
    const awarenessStates = doc.awareness.getStates();
    if (awarenessStates.size > 0) {
      const encoder2 = encoding2.createEncoder();
      encoding2.writeVarUint(encoder2, messageAwareness);
      encoding2.writeVarUint8Array(
        encoder2,
        awarenessProtocol.encodeAwarenessUpdate(
          doc.awareness,
          Array.from(awarenessStates.keys())
        )
      );
      send(doc, conn, encoding2.toUint8Array(encoder2));
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  onConnect
});
