// @tldraw/editor/dist-cjs/lib/editor/shapes/note/toolStates/Pointing.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Pointing_exports = {};
__export(Pointing_exports, {
  Pointing: () => Pointing
});
module.exports = __toCommonJS(Pointing_exports);
var import_tlschema = require("@tldraw/tlschema");
var import_NoteShapeUtil = require("../../../shapes/note/NoteShapeUtil");
var import_StateNode = require("../../../tools/StateNode");
class Pointing extends import_StateNode.StateNode {
  static id = "pointing";
  dragged = false;
  info = {};
  wasFocusedOnEnter = false;
  markPointId = "creating";
  onEnter = () => {
    this.wasFocusedOnEnter = !this.editor.isMenuOpen;
  };
  onPointerMove = (info) => {
    if (this.editor.inputs.isDragging) {
      this.editor.mark(this.markPointId);
      const shape = this.createShape();
      if (!shape)
        return;
      this.editor.setSelectedTool("select.translating", {
        ...info,
        target: "shape",
        shape,
        isCreating: true,
        editAfterComplete: true,
        onInteractionEnd: "note"
      });
    }
  };
  onPointerUp = () => {
    this.complete();
  };
  onInterrupt = () => {
    this.cancel();
  };
  onComplete = () => {
    this.complete();
  };
  onCancel = () => {
    this.cancel();
  };
  complete() {
    if (!this.wasFocusedOnEnter) {
      return;
    }
    this.editor.mark(this.markPointId);
    const shape = this.createShape();
    if (this.editor.instanceState.isToolLocked) {
      this.parent.transition("idle", {});
    } else {
      if (!shape)
        return;
      this.editor.setEditingId(shape.id);
      this.editor.setSelectedTool("select.editing_shape", {
        ...this.info,
        target: "shape",
        shape
      });
    }
  }
  cancel() {
    this.editor.bailToMark(this.markPointId);
    this.parent.transition("idle", this.info);
  }
  createShape() {
    const {
      inputs: { originPagePoint }
    } = this.editor;
    const id = (0, import_tlschema.createShapeId)();
    this.editor.createShapes(
      [
        {
          id,
          type: "note",
          x: originPagePoint.x,
          y: originPagePoint.y
        }
      ],
      true
    );
    const util = this.editor.getShapeUtil(import_NoteShapeUtil.NoteShapeUtil);
    const shape = this.editor.getShapeById(id);
    const bounds = util.bounds(shape);
    this.editor.updateShapes([
      {
        id,
        type: "note",
        x: shape.x - bounds.width / 2,
        y: shape.y - bounds.height / 2
      }
    ]);
    return this.editor.getShapeById(id);
  }
}
