// @tldraw/indices/dist-cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  getIndexAbove: () => import_reordering.getIndexAbove,
  getIndexBelow: () => import_reordering.getIndexBelow,
  getIndexBetween: () => import_reordering.getIndexBetween,
  getIndices: () => import_reordering.getIndices,
  getIndicesAbove: () => import_reordering.getIndicesAbove,
  getIndicesBelow: () => import_reordering.getIndicesBelow,
  getIndicesBetween: () => import_reordering.getIndicesBetween,
  sortByIndex: () => import_reordering.sortByIndex
});
module.exports = __toCommonJS(src_exports);
var import_reordering = require("./lib/reordering");
