// @tldraw/editor/dist-esm/lib/editor/shapes/frame/FrameShapeUtil.mjs
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { canolicalizeRotation, toDomPrecision } from "@tldraw/primitives";
import { last } from "@tldraw/utils";
import { SVGContainer } from "../../../components/SVGContainer.mjs";
import { defaultEmptyAs } from "../../../utils/string.mjs";
import { BaseBoxShapeUtil } from "../BaseBoxShapeUtil.mjs";
import { GroupShapeUtil } from "../group/GroupShapeUtil.mjs";
import { createTextSvgElementFromSpans } from "../shared/createTextSvgElementFromSpans.mjs";
import { FrameHeading } from "./components/FrameHeading.mjs";
class FrameShapeUtil extends BaseBoxShapeUtil {
  static type = "frame";
  canBind = () => true;
  canEdit = () => true;
  defaultProps() {
    return { w: 160 * 2, h: 90 * 2, name: "" };
  }
  component(shape) {
    const bounds = this.bounds(shape);
    return /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsxs(SVGContainer, { children: [
        /* @__PURE__ */ jsx("rect", { className: "tl-hitarea-stroke", width: bounds.width, height: bounds.height }),
        /* @__PURE__ */ jsx(
          "rect",
          {
            className: "tl-frame__body",
            width: bounds.width,
            height: bounds.height,
            fill: "none"
          }
        )
      ] }),
      /* @__PURE__ */ jsx(
        FrameHeading,
        {
          id: shape.id,
          name: shape.props.name,
          width: bounds.width,
          height: bounds.height
        }
      )
    ] });
  }
  toSvg(shape, font, colors) {
    const g = document.createElementNS("http://www.w3.org/2000/svg", "g");
    const rect = document.createElementNS("http://www.w3.org/2000/svg", "rect");
    rect.setAttribute("width", shape.props.w.toString());
    rect.setAttribute("height", shape.props.h.toString());
    rect.setAttribute("fill", colors.solid);
    rect.setAttribute("stroke", colors.fill.black);
    rect.setAttribute("stroke-width", "1");
    rect.setAttribute("rx", "1");
    rect.setAttribute("ry", "1");
    g.appendChild(rect);
    const pageRotation = canolicalizeRotation(this.editor.getPageRotationById(shape.id));
    const offsetRotation = pageRotation + Math.PI / 4;
    const scaledRotation = (offsetRotation * (2 / Math.PI) + 4) % 4;
    const labelSide = ["top", "left", "bottom", "right"][Math.floor(scaledRotation)];
    let labelTranslate;
    switch (labelSide) {
      case "top":
        labelTranslate = ``;
        break;
      case "right":
        labelTranslate = `translate(${toDomPrecision(shape.props.w)}px, 0px) rotate(90deg)`;
        break;
      case "bottom":
        labelTranslate = `translate(${toDomPrecision(shape.props.w)}px, ${toDomPrecision(
          shape.props.h
        )}px) rotate(180deg)`;
        break;
      case "left":
        labelTranslate = `translate(0px, ${toDomPrecision(shape.props.h)}px) rotate(270deg)`;
        break;
      default:
        labelTranslate = ``;
    }
    const opts = {
      fontSize: 12,
      fontFamily: "Inter, sans-serif",
      textAlign: "start",
      width: shape.props.w,
      height: 32,
      padding: 0,
      lineHeight: 1,
      fontStyle: "normal",
      fontWeight: "normal",
      overflow: "truncate-ellipsis",
      verticalTextAlign: "middle"
    };
    const spans = this.editor.textMeasure.measureTextSpans(
      defaultEmptyAs(shape.props.name, "Frame") + String.fromCharCode(8203),
      opts
    );
    const firstSpan = spans[0];
    const lastSpan = last(spans);
    const labelTextWidth = lastSpan.box.w + lastSpan.box.x - firstSpan.box.x;
    const text = createTextSvgElementFromSpans(this.editor, spans, {
      offsetY: -opts.height - 2,
      ...opts
    });
    text.style.setProperty("transform", labelTranslate);
    const textBg = document.createElementNS("http://www.w3.org/2000/svg", "rect");
    textBg.setAttribute("x", "-8px");
    textBg.setAttribute("y", -opts.height - 4 + "px");
    textBg.setAttribute("width", labelTextWidth + 16 + "px");
    textBg.setAttribute("height", `${opts.height}px`);
    textBg.setAttribute("rx", "4px");
    textBg.setAttribute("ry", "4px");
    textBg.setAttribute("fill", colors.background);
    g.appendChild(textBg);
    g.appendChild(text);
    return g;
  }
  indicator(shape) {
    const bounds = this.bounds(shape);
    return /* @__PURE__ */ jsx(
      "rect",
      {
        width: toDomPrecision(bounds.width),
        height: toDomPrecision(bounds.height),
        className: `tl-frame-indicator`
      }
    );
  }
  canReceiveNewChildrenOfType = (shape, _type) => {
    return !shape.isLocked;
  };
  providesBackgroundForChildren() {
    return true;
  }
  canDropShapes = (shape, _shapes) => {
    return !shape.isLocked;
  };
  onDragShapesOver = (frame, shapes) => {
    if (!shapes.every((child) => child.parentId === frame.id)) {
      this.editor.reparentShapesById(
        shapes.map((shape) => shape.id),
        frame.id
      );
      return { shouldHint: true };
    }
    return { shouldHint: false };
  };
  onDragShapesOut = (_shape, shapes) => {
    const parent = this.editor.getShapeById(_shape.parentId);
    const isInGroup = parent && this.editor.isShapeOfType(parent, GroupShapeUtil);
    if (isInGroup) {
      this.editor.reparentShapesById(
        shapes.map((shape) => shape.id),
        parent.id
      );
    } else {
      this.editor.reparentShapesById(
        shapes.map((shape) => shape.id),
        this.editor.currentPageId
      );
    }
  };
  onResizeEnd = (shape) => {
    const bounds = this.editor.getPageBounds(shape);
    const children = this.editor.getSortedChildIds(shape.id);
    const shapesToReparent = [];
    for (const childId of children) {
      const childBounds = this.editor.getPageBoundsById(childId);
      if (!bounds.includes(childBounds)) {
        shapesToReparent.push(childId);
      }
    }
    if (shapesToReparent.length > 0) {
      this.editor.reparentShapesById(shapesToReparent, this.editor.currentPageId);
    }
  };
}
export {
  FrameShapeUtil
};
