// @tldraw/editor/dist-esm/lib/constants.mjs
import { EASINGS } from "@tldraw/primitives";
const MAX_SHAPES_PER_PAGE = 2e3;
const MAX_PAGES = 40;
const REMOVE_SYMBOL = Symbol("remove");
const RICH_TYPES = {
  Date: true,
  RegExp: true,
  String: true,
  Number: true
};
const ANIMATION_SHORT_MS = 80;
const ANIMATION_MEDIUM_MS = 320;
const ZOOMS = [0.1, 0.25, 0.5, 1, 2, 4, 8];
const MIN_ZOOM = 0.1;
const MAX_ZOOM = 8;
const FOLLOW_CHASE_PROPORTION = 0.5;
const FOLLOW_CHASE_PAN_SNAP = 0.1;
const FOLLOW_CHASE_PAN_UNSNAP = 0.2;
const FOLLOW_CHASE_ZOOM_SNAP = 5e-3;
const FOLLOW_CHASE_ZOOM_UNSNAP = 0.05;
const MAJOR_NUDGE_FACTOR = 10;
const MINOR_NUDGE_FACTOR = 1;
const MAX_ASSET_WIDTH = 1e3;
const MAX_ASSET_HEIGHT = 1e3;
const GRID_INCREMENT = 5;
const MIN_CROP_SIZE = 8;
const DOUBLE_CLICK_DURATION = 450;
const MULTI_CLICK_DURATION = 200;
const COARSE_DRAG_DISTANCE = 6;
const DRAG_DISTANCE = 4;
const SVG_PADDING = 32;
const HASH_PATERN_ZOOM_NAMES = {};
for (let zoom = 1; zoom <= Math.ceil(MAX_ZOOM); zoom++) {
  HASH_PATERN_ZOOM_NAMES[zoom + "_dark"] = `hash_pattern_zoom_${zoom}_dark`;
  HASH_PATERN_ZOOM_NAMES[zoom + "_light"] = `hash_pattern_zoom_${zoom}_light`;
}
const DEFAULT_ANIMATION_OPTIONS = {
  duration: 0,
  easing: EASINGS.easeInOutCubic
};
const HAND_TOOL_FRICTION = 0.09;
const GRID_STEPS = [
  { min: -1, mid: 0.15, step: 100 },
  { min: 0.05, mid: 0.375, step: 25 },
  { min: 0.15, mid: 1, step: 5 },
  { min: 0.7, mid: 2.5, step: 1 }
];
const COLLABORATOR_TIMEOUT = 3e3;
const COLLABORATOR_CHECK_INTERVAL = 1200;
const INTERNAL_POINTER_IDS = {
  CAMERA_MOVE: -10
};
const CAMERA_MOVING_TIMEOUT = 64;
const CAMERA_MAX_RENDERING_INTERVAL = 620;
export {
  ANIMATION_MEDIUM_MS,
  ANIMATION_SHORT_MS,
  CAMERA_MAX_RENDERING_INTERVAL,
  CAMERA_MOVING_TIMEOUT,
  COARSE_DRAG_DISTANCE,
  COLLABORATOR_CHECK_INTERVAL,
  COLLABORATOR_TIMEOUT,
  DEFAULT_ANIMATION_OPTIONS,
  DOUBLE_CLICK_DURATION,
  DRAG_DISTANCE,
  FOLLOW_CHASE_PAN_SNAP,
  FOLLOW_CHASE_PAN_UNSNAP,
  FOLLOW_CHASE_PROPORTION,
  FOLLOW_CHASE_ZOOM_SNAP,
  FOLLOW_CHASE_ZOOM_UNSNAP,
  GRID_INCREMENT,
  GRID_STEPS,
  HAND_TOOL_FRICTION,
  HASH_PATERN_ZOOM_NAMES,
  INTERNAL_POINTER_IDS,
  MAJOR_NUDGE_FACTOR,
  MAX_ASSET_HEIGHT,
  MAX_ASSET_WIDTH,
  MAX_PAGES,
  MAX_SHAPES_PER_PAGE,
  MAX_ZOOM,
  MINOR_NUDGE_FACTOR,
  MIN_CROP_SIZE,
  MIN_ZOOM,
  MULTI_CLICK_DURATION,
  REMOVE_SYMBOL,
  RICH_TYPES,
  SVG_PADDING,
  ZOOMS
};
