// @tldraw/editor/dist-esm/lib/editor/shapes/geo/components/DrawStylePolygon.mjs
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { getRoundedInkyPolygonPath, getRoundedPolygonPoints } from "@tldraw/primitives";
import * as React from "react";
import { getShapeFillSvg, getSvgWithShapeFill, ShapeFill } from "../../shared/ShapeFill.mjs";
const DrawStylePolygon = React.memo(function DrawStylePolygon2({
  id,
  outline,
  lines,
  fill,
  color,
  strokeWidth
}) {
  const polygonPoints = getRoundedPolygonPoints(id, outline, strokeWidth / 3, strokeWidth * 2, 2);
  let strokePathData = getRoundedInkyPolygonPath(polygonPoints);
  if (lines) {
    for (const [A, B] of lines) {
      strokePathData += `M${A.x},${A.y}L${B.x},${B.y}`;
    }
  }
  const innerPolygonPoints = getRoundedPolygonPoints(id, outline, 0, strokeWidth * 2, 1);
  const innerPathData = getRoundedInkyPolygonPath(innerPolygonPoints);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(ShapeFill, { d: innerPathData, fill, color }),
    /* @__PURE__ */ jsx(
      "path",
      {
        d: strokePathData,
        stroke: `var(--palette-${color})`,
        strokeWidth,
        fill: "none"
      }
    )
  ] });
});
function DrawStylePolygonSvg({
  id,
  outline,
  lines,
  fill,
  color,
  colors,
  strokeWidth
}) {
  const polygonPoints = getRoundedPolygonPoints(id, outline, strokeWidth / 3, strokeWidth * 2, 2);
  let strokePathData = getRoundedInkyPolygonPath(polygonPoints);
  if (lines) {
    for (const [A, B] of lines) {
      strokePathData += `M${A.x},${A.y}L${B.x},${B.y}`;
    }
  }
  const innerPolygonPoints = getRoundedPolygonPoints(id, outline, 0, strokeWidth * 2, 1);
  const innerPathData = getRoundedInkyPolygonPath(innerPolygonPoints);
  const strokeElement = document.createElementNS("http://www.w3.org/2000/svg", "path");
  strokeElement.setAttribute("d", strokePathData);
  strokeElement.setAttribute("fill", "none");
  strokeElement.setAttribute("stroke", colors.fill[color]);
  strokeElement.setAttribute("stroke-width", strokeWidth.toString());
  const fillElement = getShapeFillSvg({
    d: innerPathData,
    fill,
    color,
    colors
  });
  return getSvgWithShapeFill(strokeElement, fillElement);
}
export {
  DrawStylePolygon,
  DrawStylePolygonSvg
};
