// @tldraw/editor/dist-cjs/lib/editor/shapes/arrow/toolStates/Pointing.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Pointing_exports = {};
__export(Pointing_exports, {
  Pointing: () => Pointing
});
module.exports = __toCommonJS(Pointing_exports);
var import_tlschema = require("@tldraw/tlschema");
var import_ArrowShapeUtil = require("../../../shapes/arrow/ArrowShapeUtil");
var import_StateNode = require("../../../tools/StateNode");
class Pointing extends import_StateNode.StateNode {
  static id = "pointing";
  shape;
  preciseTimeout = -1;
  didTimeout = false;
  startPreciseTimeout() {
    this.preciseTimeout = window.setTimeout(() => {
      if (!this.isActive)
        return;
      if (!this.shape)
        return;
      this.didTimeout = true;
    }, 300);
  }
  clearPreciseTimeout() {
    clearTimeout(this.preciseTimeout);
  }
  onEnter = () => {
    const {
      inputs: { currentPagePoint }
    } = this.editor;
    this.didTimeout = false;
    this.editor.mark("creating");
    const id = (0, import_tlschema.createShapeId)();
    this.editor.createShapes([
      {
        id,
        type: "arrow",
        x: currentPagePoint.x,
        y: currentPagePoint.y
      }
    ]);
    const util = this.editor.getShapeUtil(import_ArrowShapeUtil.ArrowShapeUtil);
    const shape = this.editor.getShapeById(id);
    if (!shape)
      return;
    const handles = util.handles?.(shape);
    if (handles) {
      const point = this.editor.getPointInShapeSpace(shape, currentPagePoint);
      const change = util.onHandleChange?.(shape, {
        handle: { ...handles[0], x: point.x, y: point.y },
        isPrecise: true
      });
      if (change) {
        const startTerminal = change.props?.start;
        if (startTerminal?.type === "binding") {
          this.editor.setHintingIds([startTerminal.boundShapeId]);
        }
        this.editor.updateShapes([change], true);
      }
    }
    this.editor.select(id);
    this.shape = this.editor.getShapeById(id);
    this.startPreciseTimeout();
  };
  onExit = () => {
    this.clearPreciseTimeout();
  };
  onPointerMove = () => {
    if (!this.shape)
      return;
    if (this.editor.inputs.isDragging) {
      const util = this.editor.getShapeUtil(this.shape);
      const handles = util.handles?.(this.shape);
      if (!handles) {
        this.editor.bailToMark("creating");
        throw Error("No handles found");
      }
      if (!this.didTimeout) {
        const util2 = this.editor.getShapeUtil(import_ArrowShapeUtil.ArrowShapeUtil);
        const shape = this.editor.getShapeById(this.shape.id);
        if (!shape)
          return;
        const handles2 = util2.handles(shape);
        if (handles2) {
          const { x, y } = this.editor.getPointInShapeSpace(
            shape,
            this.editor.inputs.originPagePoint
          );
          const change = util2.onHandleChange?.(shape, {
            handle: {
              ...handles2[0],
              x,
              y
            },
            isPrecise: false
          });
          if (change) {
            this.editor.updateShapes([change], true);
          }
        }
      }
      this.editor.setSelectedTool("select.dragging_handle", {
        shape: this.shape,
        handle: handles.find((h) => h.id === "end"),
        isCreating: true,
        onInteractionEnd: "arrow"
      });
    }
  };
  onPointerUp = () => {
    this.cancel();
  };
  onCancel = () => {
    this.cancel();
  };
  onComplete = () => {
    this.cancel();
  };
  onInterrupt = () => {
    this.cancel();
  };
  cancel() {
    this.editor.bailToMark("creating");
    this.editor.setHintingIds([]);
    this.parent.transition("idle", {});
  }
}
