// @tldraw/utils/dist-esm/index.mjs
import { compact, dedupe, last, minBy, partition, rotateArray } from "./lib/array.mjs";
import {
  Result,
  assert,
  assertExists,
  exhaustiveSwitchError,
  promiseWithResolve
} from "./lib/control.mjs";
import { debounce } from "./lib/debounce.mjs";
import { annotateError, getErrorAnnotations } from "./lib/error.mjs";
import { noop, omitFromStackTrace, throttle } from "./lib/function.mjs";
import { getHashForObject, getHashForString, lns } from "./lib/hash.mjs";
import { getFirstFromIterable } from "./lib/iterable.mjs";
import { lerp, modulate, rng } from "./lib/number.mjs";
import {
  deepCopy,
  filterEntries,
  getOwnProperty,
  hasOwnProperty,
  mapObjectMapValues,
  objectMapEntries,
  objectMapFromEntries,
  objectMapKeys,
  objectMapValues
} from "./lib/object.mjs";
import { rafThrottle, throttledRaf } from "./lib/raf.mjs";
import { sortById } from "./lib/sort.mjs";
import { isDefined, isNonNull, isNonNullish, structuredClone } from "./lib/value.mjs";
export {
  Result,
  annotateError,
  assert,
  assertExists,
  compact,
  debounce,
  dedupe,
  deepCopy,
  exhaustiveSwitchError,
  filterEntries,
  getErrorAnnotations,
  getFirstFromIterable,
  getHashForObject,
  getHashForString,
  getOwnProperty,
  hasOwnProperty,
  isDefined,
  isNonNull,
  isNonNullish,
  last,
  lerp,
  lns,
  mapObjectMapValues,
  minBy,
  modulate,
  noop,
  objectMapEntries,
  objectMapFromEntries,
  objectMapKeys,
  objectMapValues,
  omitFromStackTrace,
  partition,
  promiseWithResolve,
  rafThrottle,
  rng,
  rotateArray,
  sortById,
  structuredClone,
  throttle,
  throttledRaf
};
