// @tldraw/primitives/dist-esm/lib/LineSegment2d.mjs
import { BaseSegment2d } from "./BaseSegment2d.mjs";
import { Box2d } from "./Box2d.mjs";
import { Vec2d } from "./Vec2d.mjs";
class LineSegment2d extends BaseSegment2d {
  constructor(a, b, p = 2) {
    super({ a, b, p });
    this.a = a;
    this.b = b;
    this.values.p = Math.max(10, Math.ceil(this.length / 20));
  }
  get length() {
    return Vec2d.Dist(this.a, this.b);
  }
  get tangent() {
    return Vec2d.Tan(this.a, this.b);
  }
  get angle() {
    return Vec2d.Angle(this.a, this.b);
  }
  get bounds() {
    return Box2d.FromPoints([this.a, this.b]);
  }
  getX(t) {
    const { a, b } = this.values;
    return a.x + (b.x - a.x) * t;
  }
  getY(t) {
    const { a, b } = this.values;
    return a.y + (b.y - a.y) * t;
  }
  getPoint(t) {
    const { a, b } = this.values;
    if (t <= 0)
      return Vec2d.From(a);
    if (t >= 1)
      return Vec2d.From(b);
    return Vec2d.Lrp(a, b, t);
  }
  getPath(head = true) {
    const { a, b } = this.values;
    if (Vec2d.Equals(a, b))
      return "";
    return head ? `M${a.x},${a.y}L${b.x},${b.y}` : `${b.x},${b.y}`;
  }
  getNormal() {
    const { a, b } = this.values;
    return Vec2d.Sub(a, b).per().uni().toFixed();
  }
  /**
   * Get the closest point on the segment to an arbitrary point.
   *
   * @param point - The arbitrary point.
   * @public
   */
  getClosestPointTo(point) {
    const { a, b } = this.values;
    const closestPoint = Vec2d.NearestPointOnLineSegment(a, b, point);
    const closestDistance = closestPoint.dist(point);
    return { point: closestPoint, distance: closestDistance };
  }
  static Length(A) {
    return Vec2d.Dist(A.a, A.b);
  }
  static Tangent(A) {
    return Vec2d.Tan(A.a, A.b);
  }
  static Angle(A) {
    return Vec2d.Angle(A.a, A.b);
  }
}
export {
  LineSegment2d
};
